// @taiga-ui/legacy/esm2022/components/primitive-textfield/primitive-textfield.directive.mjs
import { Directive, inject } from '@angular/core';
import { AbstractTuiTextfieldHost } from '@taiga-ui/legacy/classes';
import { tuiAsTextfieldHost } from '@taiga-ui/legacy/tokens';
import { TuiPrimitiveTextfieldComponent } from './primitive-textfield.component';
import * as i0 from "@angular/core";
/**
 * @deprecated: use {@link TuiTextfield}
 * https://taiga-ui.dev/components/textfield
 */
class TuiPrimitiveTextfieldDirective extends AbstractTuiTextfieldHost {
    constructor() {
        super(...arguments);
        this.host = inject(TuiPrimitiveTextfieldComponent);
    }
    get readOnly() {
        return this.host.readOnly || !this.host.editable;
    }
    onValueChange(value) {
        this.host.onModelChange(value);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPrimitiveTextfieldDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiPrimitiveTextfieldDirective, selector: "tui-primitive-textfield", providers: [tuiAsTextfieldHost(TuiPrimitiveTextfieldDirective)], usesInheritance: true, ngImport: i0 }); }
}
export { TuiPrimitiveTextfieldDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPrimitiveTextfieldDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: false,
                    selector: 'tui-primitive-textfield',
                    providers: [tuiAsTextfieldHost(TuiPrimitiveTextfieldDirective)],
                }]
        }] });
