// @taiga-ui/legacy/esm2022/components/primitive-textfield/primitive-textfield.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiIcon } from '@taiga-ui/core/components/icon';
import { TuiAppearance } from '@taiga-ui/core/directives/appearance';
import { TuiTooltipModule } from '@taiga-ui/legacy/components/tooltip';
import { TuiWrapperModule } from '@taiga-ui/legacy/directives/wrapper';
import { PolymorpheusOutlet, PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import { TuiPrimitiveTextfieldComponent } from './primitive-textfield.component';
import { TuiPrimitiveTextfieldDirective } from './primitive-textfield.directive';
import { TuiTextfieldComponent } from './textfield/textfield.component';
import { TuiValueDecorationComponent } from './value-decoration/value-decoration.component';
import * as i0 from "@angular/core";
/**
 * @deprecated: use {@link TuiTextfield}
 * https://taiga-ui.dev/components/textfield
 */
class TuiPrimitiveTextfieldModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPrimitiveTextfieldModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiPrimitiveTextfieldModule, declarations: [TuiPrimitiveTextfieldComponent,
            TuiPrimitiveTextfieldDirective,
            TuiTextfieldComponent,
            TuiValueDecorationComponent], imports: [CommonModule,
            FormsModule,
            PolymorpheusOutlet,
            PolymorpheusTemplate,
            TuiWrapperModule,
            TuiTooltipModule,
            TuiIcon,
            TuiAppearance], exports: [TuiPrimitiveTextfieldComponent,
            TuiPrimitiveTextfieldDirective,
            TuiTextfieldComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPrimitiveTextfieldModule, imports: [CommonModule,
            FormsModule,
            TuiWrapperModule,
            TuiTooltipModule,
            TuiIcon] }); }
}
export { TuiPrimitiveTextfieldModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPrimitiveTextfieldModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        PolymorpheusOutlet,
                        PolymorpheusTemplate,
                        TuiWrapperModule,
                        TuiTooltipModule,
                        TuiIcon,
                        TuiAppearance,
                    ],
                    declarations: [
                        TuiPrimitiveTextfieldComponent,
                        TuiPrimitiveTextfieldDirective,
                        TuiTextfieldComponent,
                        TuiValueDecorationComponent,
                    ],
                    exports: [
                        TuiPrimitiveTextfieldComponent,
                        TuiPrimitiveTextfieldDirective,
                        TuiTextfieldComponent,
                    ],
                }]
        }] });
