// @taiga-ui/legacy/esm2022/components/input-copy/input-copy.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiIcon } from '@taiga-ui/core/components/icon';
import { TuiAppearance } from '@taiga-ui/core/directives/appearance';
import { TuiHint } from '@taiga-ui/core/directives/hint';
import { TuiPrimitiveTextfieldModule, TuiTextfieldComponent, } from '@taiga-ui/legacy/components/primitive-textfield';
import { TuiTextfieldControllerModule } from '@taiga-ui/legacy/directives';
import { PolymorpheusOutlet, PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import { TuiInputCopyComponent } from './input-copy.component';
import { TuiInputCopyDirective } from './input-copy.directive';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/hint";
/**
 * @deprecated use {@link TuiCopy} with {@link TuiTextfield}
 */
class TuiInputCopyModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputCopyModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiInputCopyModule, declarations: [TuiInputCopyComponent, TuiInputCopyDirective], imports: [CommonModule,
            PolymorpheusOutlet,
            PolymorpheusTemplate,
            TuiPrimitiveTextfieldModule,
            TuiTextfieldControllerModule,
            TuiAppearance,
            TuiIcon, i1.TuiHintComponent, i1.TuiHintDirective, i1.TuiHintOptionsDirective, i1.TuiHintUnstyled, i1.TuiHintDriver, i1.TuiHintPosition, i1.TuiHintHover, i1.TuiHintOverflow, i1.TuiHintDescribe, i1.TuiHintHost, i1.TuiHintManual, i1.TuiHintPointer], exports: [TuiInputCopyComponent,
            TuiInputCopyDirective,
            TuiTextfieldComponent, i1.TuiHintComponent, i1.TuiHintDirective, i1.TuiHintOptionsDirective, i1.TuiHintUnstyled, i1.TuiHintDriver, i1.TuiHintPosition, i1.TuiHintHover, i1.TuiHintOverflow, i1.TuiHintDescribe, i1.TuiHintHost, i1.TuiHintManual, i1.TuiHintPointer] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputCopyModule, imports: [CommonModule,
            TuiPrimitiveTextfieldModule,
            TuiTextfieldControllerModule,
            TuiIcon, i1.TuiHintComponent] }); }
}
export { TuiInputCopyModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputCopyModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PolymorpheusOutlet,
                        PolymorpheusTemplate,
                        TuiPrimitiveTextfieldModule,
                        TuiTextfieldControllerModule,
                        TuiAppearance,
                        TuiIcon,
                        ...TuiHint,
                    ],
                    declarations: [TuiInputCopyComponent, TuiInputCopyDirective],
                    exports: [
                        TuiInputCopyComponent,
                        TuiInputCopyDirective,
                        TuiTextfieldComponent,
                        ...TuiHint,
                    ],
                }]
        }] });
