// @taiga-ui/cdk/schematics/ng-update/v4/steps/templates/toggles/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.closeStartTag = exports.removeClosingTag = exports.replaceSizeAttr = exports.replaceOpenTag = void 0;
const inputs_1 = require("../../../../../utils/templates/inputs");
const sizeMap = {
    l: 'm',
    m: 's',
};
function replaceOpenTag(sourceCodeLocation, recorder, templateOffset, { tag, directive, type }) {
    const { startTag } = sourceCodeLocation;
    if (!startTag) {
        return;
    }
    const spaces = ' '.repeat(startTag.startCol + 3);
    recorder.remove(templateOffset + startTag.startOffset, `<${tag}`.length);
    recorder.insertRight(templateOffset + startTag.startOffset, `<input\n${spaces}${directive}\n${spaces}type="${type}"`);
}
exports.replaceOpenTag = replaceOpenTag;
function replaceSizeAttr(attrs, sourceCodeLocation, recorder, templateOffset, map = sizeMap) {
    var _a;
    const sizeAttr = (0, inputs_1.findAttr)(attrs, 'size');
    if (sizeAttr) {
        const { startOffset, endOffset } = ((_a = sourceCodeLocation.attrs) === null || _a === void 0 ? void 0 : _a[sizeAttr.name]) || {
            startOffset: 0,
            endOffset: 0,
        };
        recorder.remove(templateOffset + startOffset, endOffset - startOffset);
        recorder.insertRight(templateOffset + startOffset, `${sizeAttr.name}="${map[sizeAttr.value] || sizeAttr.value}"`);
    }
}
exports.replaceSizeAttr = replaceSizeAttr;
function removeClosingTag(sourceCodeLocation, recorder, templateOffset) {
    const { endTag } = sourceCodeLocation;
    if (!endTag) {
        return;
    }
    const { startOffset, endOffset } = endTag;
    const from = templateOffset + startOffset;
    const to = endOffset - startOffset;
    recorder.remove(from, to);
}
exports.removeClosingTag = removeClosingTag;
function closeStartTag({ startTag, endTag }, recorder, templateOffset) {
    var _a;
    if (!endTag) {
        return;
    }
    recorder.insertRight(templateOffset + ((_a = startTag === null || startTag === void 0 ? void 0 : startTag.endOffset) !== null && _a !== void 0 ? _a : 1) - 1, '/');
}
exports.closeStartTag = closeStartTag;
