// @taiga-ui/kit/esm2022/pipes/emails/emails.pipe.mjs
import { inject, Pipe } from '@angular/core';
import { TUI_EMAIL_PIPE_OPTIONS } from './emails.options';
import * as i0 from "@angular/core";
class TuiEmailsPipe {
    constructor() {
        this.options = inject(TUI_EMAIL_PIPE_OPTIONS);
    }
    transform(query, suggestions = this.options) {
        return query.includes('@')
            ? suggestions
                .map((item) => query.slice(0, Math.max(0, query.indexOf('@') + 1)) + item)
                .filter((item) => item.startsWith(query))
            : [];
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiEmailsPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiEmailsPipe, isStandalone: true, name: "tuiEmails" }); }
}
export { TuiEmailsPipe };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiEmailsPipe, decorators: [{
            type: Pipe,
            args: [{
                    standalone: true,
                    name: 'tuiEmails',
                }]
        }] });
