// @taiga-ui/legacy/esm2022/components/input-slider/input-slider.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiNumberFormat } from '@taiga-ui/core/directives/number-format';
import { TuiSlider } from '@taiga-ui/kit/components/slider';
import { TuiInputNumberModule } from '@taiga-ui/legacy/components/input-number';
import { TuiTextfieldControllerModule } from '@taiga-ui/legacy/directives';
import { PolymorpheusOutlet, PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import { TuiInputSliderComponent } from './input-slider.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/kit/components/slider";
class TuiInputSliderModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputSliderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiInputSliderModule, declarations: [TuiInputSliderComponent], imports: [CommonModule,
            FormsModule,
            PolymorpheusOutlet,
            PolymorpheusTemplate,
            TuiInputNumberModule,
            TuiNumberFormat,
            TuiTextfieldControllerModule, i1.TuiSliderComponent, i1.TuiSliderThumbLabel, i1.TuiSliderKeySteps, i1.TuiSliderReadonly], exports: [TuiInputSliderComponent, i1.TuiSliderComponent, i1.TuiSliderThumbLabel, i1.TuiSliderKeySteps, i1.TuiSliderReadonly] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputSliderModule, imports: [CommonModule,
            FormsModule,
            TuiInputNumberModule,
            TuiTextfieldControllerModule, i1.TuiSliderComponent, i1.TuiSliderThumbLabel] }); }
}
export { TuiInputSliderModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputSliderModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        PolymorpheusOutlet,
                        PolymorpheusTemplate,
                        TuiInputNumberModule,
                        TuiNumberFormat,
                        TuiTextfieldControllerModule,
                        ...TuiSlider,
                    ],
                    declarations: [TuiInputSliderComponent],
                    exports: [TuiInputSliderComponent, ...TuiSlider],
                }]
        }] });
