// @taiga-ui/kit/esm2022/components/progress/progress-bar/progress-color-segments.directive.mjs
import { ChangeDetectorRef, Directive, inject, Input } from '@angular/core';
import { toSignal } from '@angular/core/rxjs-interop';
import { ResizeObserverService } from '@ng-web-apis/resize-observer';
import { tuiWatch, tuiZonefull } from '@taiga-ui/cdk/observables';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { BehaviorSubject, combineLatest, distinctUntilChanged, map } from 'rxjs';
import * as i0 from "@angular/core";
class TuiProgressColorSegments {
    constructor() {
        this.colors$ = new BehaviorSubject([]);
        this.el = tuiInjectElement();
        this.color = toSignal(combineLatest([
            this.colors$,
            inject(ResizeObserverService, { self: true }).pipe(map(() => this.el.offsetWidth), distinctUntilChanged()),
        ]).pipe(map(([colors, width]) => {
            const segmentWidth = Math.ceil(width / colors.length);
            const colorsString = colors.reduce((acc, color, i) => `${acc}, ${color} ${i * segmentWidth}px ${(i + 1) * segmentWidth}px`, '');
            return `linear-gradient(to right ${colorsString})`;
        }), tuiZonefull(), tuiWatch(inject(ChangeDetectorRef))));
    }
    set colors(colors) {
        this.colors$.next(colors);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiProgressColorSegments, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiProgressColorSegments, isStandalone: true, selector: "progress[tuiProgressBar][tuiProgressColorSegments]", inputs: { colors: ["tuiProgressColorSegments", "colors"] }, host: { properties: { "style.--tui-progress-color": "color()" } }, providers: [ResizeObserverService], ngImport: i0 }); }
}
export { TuiProgressColorSegments };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiProgressColorSegments, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'progress[tuiProgressBar][tuiProgressColorSegments]',
                    providers: [ResizeObserverService],
                    host: { '[style.--tui-progress-color]': 'color()' },
                }]
        }], propDecorators: { colors: [{
                type: Input,
                args: ['tuiProgressColorSegments']
            }] } });
