// @taiga-ui/kit/esm2022/directives/button-close/button-close.directive.mjs
import { Directive, inject } from '@angular/core';
import { tuiButtonOptionsProvider } from '@taiga-ui/core/components/button';
import { TUI_COMMON_ICONS, TUI_ICON_START } from '@taiga-ui/core/tokens';
import * as i0 from "@angular/core";
class TuiButtonClose {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiButtonClose, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiButtonClose, isStandalone: true, selector: "[tuiIconButton][tuiButtonClose]", host: { properties: { "style.--t-radius.%": "100", "style.--tui-height-s.rem": "1.875" } }, providers: [
            tuiButtonOptionsProvider({ appearance: 'neutral', size: 's' }),
            {
                provide: TUI_ICON_START,
                useFactory: () => inject(TUI_COMMON_ICONS).close,
            },
        ], ngImport: i0 }); }
}
export { TuiButtonClose };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiButtonClose, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiIconButton][tuiButtonClose]',
                    providers: [
                        tuiButtonOptionsProvider({ appearance: 'neutral', size: 's' }),
                        {
                            provide: TUI_ICON_START,
                            useFactory: () => inject(TUI_COMMON_ICONS).close,
                        },
                    ],
                    host: {
                        '[style.--t-radius.%]': '100',
                        '[style.--tui-height-s.rem]': '1.875',
                    },
                }]
        }] });
