// @taiga-ui/legacy/esm2022/components/input-date-range/input-date-range.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MaskitoDirective } from '@maskito/angular';
import { TuiLet } from '@taiga-ui/cdk/directives/let';
import { TuiIcon } from '@taiga-ui/core/components/icon';
import { TuiAppearance } from '@taiga-ui/core/directives/appearance';
import { TuiDropdown } from '@taiga-ui/core/directives/dropdown';
import { TuiCalendarRange } from '@taiga-ui/kit/components/calendar-range';
import { TuiPrimitiveTextfieldModule, TuiTextfieldComponent, } from '@taiga-ui/legacy/components/primitive-textfield';
import { TuiLegacyDropdownOpenMonitorDirective, TuiTextfieldControllerModule, TuiValueAccessorModule, } from '@taiga-ui/legacy/directives';
import { PolymorpheusOutlet, PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import { TuiInputDateRangeComponent } from './input-date-range.component';
import { TuiInputDateRangeDirective } from './input-date-range.directive';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/dropdown";
class TuiInputDateRangeModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateRangeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateRangeModule, declarations: [TuiInputDateRangeComponent, TuiInputDateRangeDirective], imports: [CommonModule,
            MaskitoDirective,
            TuiLet,
            PolymorpheusOutlet,
            PolymorpheusTemplate,
            TuiPrimitiveTextfieldModule,
            TuiTextfieldControllerModule,
            TuiCalendarRange,
            TuiValueAccessorModule,
            TuiLegacyDropdownOpenMonitorDirective,
            TuiIcon,
            TuiAppearance, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection], exports: [TuiInputDateRangeComponent,
            TuiInputDateRangeDirective,
            TuiTextfieldComponent, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateRangeModule, imports: [CommonModule,
            TuiPrimitiveTextfieldModule,
            TuiTextfieldControllerModule,
            TuiCalendarRange,
            TuiValueAccessorModule,
            TuiIcon, i1.TuiDropdownComponent] }); }
}
export { TuiInputDateRangeModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateRangeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MaskitoDirective,
                        TuiLet,
                        PolymorpheusOutlet,
                        PolymorpheusTemplate,
                        TuiPrimitiveTextfieldModule,
                        TuiTextfieldControllerModule,
                        TuiCalendarRange,
                        TuiValueAccessorModule,
                        TuiLegacyDropdownOpenMonitorDirective,
                        TuiIcon,
                        TuiAppearance,
                        ...TuiDropdown,
                    ],
                    declarations: [TuiInputDateRangeComponent, TuiInputDateRangeDirective],
                    exports: [
                        TuiInputDateRangeComponent,
                        TuiInputDateRangeDirective,
                        TuiTextfieldComponent,
                        ...TuiDropdown,
                    ],
                }]
        }] });
