// @taiga-ui/legacy/esm2022/components/sheet/directives/sheet-top/sheet-top.directive.mjs
import { Directive, inject, Input } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { WA_WINDOW } from '@ng-web-apis/common';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { tuiClamp } from '@taiga-ui/cdk/utils/math';
import { map } from 'rxjs';
import { TUI_SHEET, TUI_SHEET_SCROLL } from '../../sheet-tokens';
import * as i0 from "@angular/core";
// So that borders get rounded when image is visible for at least 10px
const OFFSET = 10;
class TuiSheetTopDirective {
    constructor() {
        this.scroll$ = inject(TUI_SHEET_SCROLL);
        this.component = inject(TUI_SHEET);
        this.win = inject(WA_WINDOW);
        this.el = tuiInjectElement();
        this.rounded$ = this.scroll$
            .pipe(map((y) => y < this.stop + OFFSET))
            .pipe(takeUntilDestroyed())
            .subscribe((add) => add
            ? this.el.classList.add('_rounded')
            : this.el.classList.remove('_rounded'));
        this.transform$ = this.scroll$
            .pipe(map((y) => `translateY(${this.getY(y)}%) scaleX(-1)`), takeUntilDestroyed())
            .subscribe((transform) => this.el.style.setProperty('transform', transform));
        this.clickthrough$ = this.scroll$
            .pipe(map((y) => !!Math.round(this.getY(y))), takeUntilDestroyed())
            .subscribe((add) => add
            ? this.el.classList.add('_clickthrough')
            : this.el.classList.remove('_clickthrough'));
        this.stop = 0;
    }
    getY(scrollTop) {
        const value = scrollTop - this.stop;
        const total = this.win.innerHeight - this.component.item.offset - this.stop;
        return this.stop && tuiClamp(100 - (value / total) * 100, 0, 100);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSheetTopDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiSheetTopDirective, selector: "[tuiSheetTop]", inputs: { stop: ["tuiSheetTop", "stop"] }, ngImport: i0 }); }
}
export { TuiSheetTopDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSheetTopDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: false,
                    selector: '[tuiSheetTop]',
                }]
        }], propDecorators: { stop: [{
                type: Input,
                args: ['tuiSheetTop']
            }] } });
