// @taiga-ui/legacy/esm2022/components/input-color/input-color.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MaskitoDirective } from '@maskito/angular';
import { TuiDropdown } from '@taiga-ui/core/directives/dropdown';
import { TuiColorSelectorModule } from '@taiga-ui/legacy/components/color-selector';
import { TuiPrimitiveTextfieldModule } from '@taiga-ui/legacy/components/primitive-textfield';
import { TuiTextfieldControllerModule } from '@taiga-ui/legacy/directives/textfield-controller';
import { TuiInputColorComponent } from './input-color.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/dropdown";
class TuiInputColorModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputColorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiInputColorModule, declarations: [TuiInputColorComponent], imports: [CommonModule,
            TuiPrimitiveTextfieldModule,
            TuiTextfieldControllerModule,
            TuiColorSelectorModule,
            MaskitoDirective, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection], exports: [TuiInputColorComponent, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputColorModule, imports: [CommonModule,
            TuiPrimitiveTextfieldModule,
            TuiTextfieldControllerModule,
            TuiColorSelectorModule, i1.TuiDropdownComponent] }); }
}
export { TuiInputColorModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputColorModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiPrimitiveTextfieldModule,
                        TuiTextfieldControllerModule,
                        TuiColorSelectorModule,
                        MaskitoDirective,
                        ...TuiDropdown,
                    ],
                    declarations: [TuiInputColorComponent],
                    exports: [TuiInputColorComponent, ...TuiDropdown],
                }]
        }] });
