// @taiga-ui/core/esm2022/services/visual-viewport.service.mjs
import { inject, Injectable } from '@angular/core';
import { WA_WINDOW } from '@ng-web-apis/common';
import { TUI_IS_WEBKIT } from '@taiga-ui/cdk/tokens';
import * as i0 from "@angular/core";
class TuiVisualViewportService {
    constructor() {
        this.isWebkit = inject(TUI_IS_WEBKIT);
        this.win = inject(WA_WINDOW);
    }
    // https://bugs.webkit.org/show_bug.cgi?id=207089
    correct(point) {
        return this.isWebkit
            ? [
                point[0] + (this.win.visualViewport?.offsetTop ?? 0),
                point[1] + (this.win.visualViewport?.offsetLeft ?? 0),
            ]
            : point;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiVisualViewportService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiVisualViewportService, providedIn: 'root' }); }
}
export { TuiVisualViewportService };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiVisualViewportService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
