// @taiga-ui/legacy/esm2022/components/sheet/sheet.service.mjs
import { DOCUMENT } from '@angular/common';
import { inject, Injectable } from '@angular/core';
import { tuiIsHTMLElement } from '@taiga-ui/cdk/utils/dom';
import { tuiGetNativeFocused } from '@taiga-ui/cdk/utils/focus';
import { BehaviorSubject, EMPTY, Observable } from 'rxjs';
import { TUI_SHEET_OPTIONS } from './sheet-options';
import * as i0 from "@angular/core";
/**
 * @deprecated: use {@link TuiSheetDialogService}
 * https://taiga-ui.dev/components/sheet-dialog
 */
class TuiSheetService {
    constructor() {
        this.doc = inject(DOCUMENT);
        this.options = inject(TUI_SHEET_OPTIONS);
        this.sheets$ = new BehaviorSubject([]);
    }
    open(content, options = {}) {
        return new Observable(($implicit) => {
            const completeWith = (result) => {
                $implicit.next(result);
                $implicit.complete();
            };
            const sheet = {
                ...this.options,
                ...options,
                content,
                completeWith,
                $implicit,
                scroll$: EMPTY,
            };
            const focused = tuiGetNativeFocused(this.doc);
            if (tuiIsHTMLElement(focused) && sheet.overlay) {
                focused.blur();
            }
            this.sheets$.next([...this.sheets$.value, sheet]);
            return () => {
                this.sheets$.next(this.sheets$.value.filter((item) => item !== sheet));
            };
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSheetService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSheetService, providedIn: 'root' }); }
}
export { TuiSheetService };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSheetService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
