// @taiga-ui/legacy/esm2022/components/sheet/sheet.directive.mjs
import { Directive, inject, Input, Output } from '@angular/core';
import { tuiIfMap } from '@taiga-ui/cdk/observables';
import { PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import { endWith, ignoreElements, share, Subject } from 'rxjs';
import { TuiSheetService } from './sheet.service';
import * as i0 from "@angular/core";
/**
 * @deprecated: use {@link TuiSheetDialog}
 * https://taiga-ui.dev/components/sheet-dialog
 */
class TuiSheetDirective extends PolymorpheusTemplate {
    constructor() {
        super(...arguments);
        this.service = inject(TuiSheetService);
        this.open$ = new Subject();
        this.options = {};
        this.tuiSheetChange = this.open$.pipe(tuiIfMap(() => this.service.open(this, this.options).pipe(ignoreElements(), endWith(false))), share());
    }
    set tuiSheet(open) {
        this.open$.next(open);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSheetDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiSheetDirective, selector: "ng-template[tuiSheet]", inputs: { options: ["tuiSheetOptions", "options"], tuiSheet: "tuiSheet" }, outputs: { tuiSheetChange: "tuiSheetChange" }, usesInheritance: true, ngImport: i0 }); }
}
export { TuiSheetDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSheetDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: false,
                    selector: 'ng-template[tuiSheet]',
                }]
        }], propDecorators: { options: [{
                type: Input,
                args: ['tuiSheetOptions']
            }], tuiSheetChange: [{
                type: Output
            }], tuiSheet: [{
                type: Input
            }] } });
