// @taiga-ui/legacy/esm2022/components/sheet/sheet.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule, NO_ERRORS_SCHEMA } from '@angular/core';
import { TuiLet } from '@taiga-ui/cdk/directives/let';
import { TuiButton } from '@taiga-ui/core/components/button';
import { TuiScrollbar } from '@taiga-ui/core/components/scrollbar';
import { PolymorpheusOutlet, PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import { TuiSheetComponent } from './components/sheet/sheet.component';
import { TuiSheetBarComponent } from './components/sheet-bar/sheet-bar.component';
import { TuiSheetHeadingComponent } from './components/sheet-heading/sheet-heading.component';
import { TuiSheetsHostComponent } from './components/sheets-host/sheets-host.component';
import { TuiSheetCloseDirective } from './directives/sheet-close/sheet-close.directive';
import { TuiSheetStopDirective } from './directives/sheet-stop/sheet-stop.directive';
import { TuiSheetTopDirective } from './directives/sheet-top/sheet-top.directive';
import { TuiSheetWrapperDirective } from './directives/sheet-wrapper/sheet-wrapper.directive';
import { TuiSheetDirective } from './sheet.directive';
import * as i0 from "@angular/core";
/**
 * @deprecated: use {@link TuiSheetDialog}
 * https://taiga-ui.dev/components/sheet-dialog
 */
class TuiSheetModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSheetModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiSheetModule, declarations: [TuiSheetsHostComponent,
            TuiSheetComponent,
            TuiSheetBarComponent,
            TuiSheetHeadingComponent,
            TuiSheetTopDirective,
            TuiSheetDirective,
            TuiSheetWrapperDirective,
            TuiSheetCloseDirective,
            TuiSheetStopDirective], imports: [CommonModule,
            PolymorpheusOutlet,
            PolymorpheusTemplate,
            TuiLet,
            TuiButton,
            TuiScrollbar], exports: [TuiSheetsHostComponent,
            TuiSheetComponent,
            TuiSheetHeadingComponent,
            TuiSheetDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSheetModule, imports: [CommonModule,
            TuiScrollbar] }); }
}
export { TuiSheetModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSheetModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PolymorpheusOutlet,
                        PolymorpheusTemplate,
                        TuiLet,
                        TuiButton,
                        TuiScrollbar,
                    ],
                    declarations: [
                        TuiSheetsHostComponent,
                        TuiSheetComponent,
                        TuiSheetBarComponent,
                        TuiSheetHeadingComponent,
                        TuiSheetTopDirective,
                        TuiSheetDirective,
                        TuiSheetWrapperDirective,
                        TuiSheetCloseDirective,
                        TuiSheetStopDirective,
                    ],
                    exports: [
                        TuiSheetsHostComponent,
                        TuiSheetComponent,
                        TuiSheetHeadingComponent,
                        TuiSheetDirective,
                    ],
                    schemas: [NO_ERRORS_SCHEMA],
                }]
        }] });
