// @taiga-ui/legacy/esm2022/components/input-date/native-date/native-date.component.mjs
import { Directive, inject } from '@angular/core';
import { DATE_FILLER_LENGTH, TuiDay } from '@taiga-ui/cdk/date-time';
import { TUI_TEXTFIELD_HOST, tuiAsTextfieldHost } from '@taiga-ui/legacy/tokens';
import { TuiInputDateDirective } from '../input-date.directive';
import * as i0 from "@angular/core";
class TuiNativeDateDirective {
    constructor() {
        this.host = inject(TUI_TEXTFIELD_HOST);
    }
    get value() {
        return this.host.value.length === DATE_FILLER_LENGTH
            ? TuiDay.normalizeParse(this.host.value, this.host.format.mode).toString('YMD', '-')
            : '';
    }
    get max() {
        return this.host.max.toJSON();
    }
    get min() {
        return this.host.min.toJSON();
    }
    onChange(value) {
        this.host.onValueChange(value
            ? TuiDay.normalizeParse(value, 'YMD').toString(this.host.format.mode)
            : '');
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiNativeDateDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiNativeDateDirective, selector: "input[tuiDate]", host: { attributes: { "type": "date" }, listeners: { "change": "onChange($event.target.value)", "click.stop.silent": "0", "input.stop.silent": "0", "mousedown.stop.silent": "0" }, properties: { "tabIndex": "-1", "value": "value", "max": "max", "min": "min" } }, providers: [tuiAsTextfieldHost(TuiInputDateDirective)], ngImport: i0 }); }
}
export { TuiNativeDateDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiNativeDateDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: false,
                    selector: 'input[tuiDate]',
                    providers: [tuiAsTextfieldHost(TuiInputDateDirective)],
                    host: {
                        type: 'date',
                        '[tabIndex]': '-1',
                        '[value]': 'value',
                        '(change)': 'onChange($event.target.value)',
                        '[max]': 'max',
                        '[min]': 'min',
                        '(click.stop.silent)': '0',
                        '(input.stop.silent)': '0',
                        '(mousedown.stop.silent)': '0',
                    },
                }]
        }] });
