// @taiga-ui/legacy/esm2022/components/multi-select-option/multi-select-option.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiCheckbox } from '@taiga-ui/kit/components/checkbox';
import { TuiMultiSelectOptionComponent } from './multi-select-option.component';
import * as i0 from "@angular/core";
class TuiMultiSelectOptionModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectOptionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectOptionModule, declarations: [TuiMultiSelectOptionComponent], imports: [CommonModule, TuiCheckbox], exports: [TuiMultiSelectOptionComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectOptionModule, imports: [CommonModule, TuiCheckbox] }); }
}
export { TuiMultiSelectOptionModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectOptionModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiCheckbox],
                    declarations: [TuiMultiSelectOptionComponent],
                    exports: [TuiMultiSelectOptionComponent],
                }]
        }] });
