// @taiga-ui/kit/esm2022/directives/lazy-loading/lazy-loading.directive.mjs
import { Directive, inject, Input } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { IntersectionObserverService } from '@ng-web-apis/intersection-observer';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { TuiLazyLoadingService } from './lazy-loading.service';
import * as i0 from "@angular/core";
class TuiImgLazyLoading {
    constructor() {
        this.el = tuiInjectElement();
        this.src$ = inject(TuiLazyLoadingService);
        this.animation = 'tuiSkeletonVibe ease-in-out 1s infinite alternate';
        this.background = 'var(--tui-background-neutral-2)';
        this.src = null;
        if (this.supported) {
            return;
        }
        this.src$.pipe(takeUntilDestroyed()).subscribe((src) => {
            this.src = src;
        });
    }
    set srcSetter(src) {
        this.src = this.supported ? src : null;
        this.src$.next(src);
    }
    onLoad() {
        this.background = '';
        this.animation = '';
    }
    get supported() {
        return 'loading' in this.el;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiImgLazyLoading, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiImgLazyLoading, isStandalone: true, selector: "img[loading=\"lazy\"]", inputs: { srcSetter: ["src", "srcSetter"] }, host: { listeners: { "load": "onLoad()" }, properties: { "style.animation": "animation", "style.background": "background", "attr.src": "src" } }, providers: [TuiLazyLoadingService, IntersectionObserverService], ngImport: i0 }); }
}
export { TuiImgLazyLoading };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiImgLazyLoading, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'img[loading="lazy"]',
                    providers: [TuiLazyLoadingService, IntersectionObserverService],
                    host: {
                        '[style.animation]': 'animation',
                        '[style.background]': 'background',
                        '[attr.src]': 'src',
                        '(load)': 'onLoad()',
                    },
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { srcSetter: [{
                type: Input,
                args: ['src']
            }] } });
