// @taiga-ui/core/esm2022/directives/hint/hint-manual.directive.mjs
import { Directive, inject, Input } from '@angular/core';
import { tuiAsDriver, TuiDriver } from '@taiga-ui/core/classes';
import { BehaviorSubject } from 'rxjs';
import { TuiHintHover } from './hint-hover.directive';
import * as i0 from "@angular/core";
class TuiHintManual extends TuiDriver {
    constructor() {
        super((subscriber) => this.stream$.subscribe(subscriber));
        this.hover = inject(TuiHintHover);
        this.stream$ = new BehaviorSubject(false);
        this.tuiHintManual = false;
        this.type = 'hint';
        this.hover.enabled = false;
    }
    ngOnChanges() {
        this.stream$.next(this.tuiHintManual);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiHintManual, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiHintManual, isStandalone: true, selector: "[tuiHint][tuiHintManual]", inputs: { tuiHintManual: "tuiHintManual" }, providers: [tuiAsDriver(TuiHintManual)], usesInheritance: true, usesOnChanges: true, ngImport: i0 }); }
}
export { TuiHintManual };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiHintManual, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiHint][tuiHintManual]',
                    providers: [tuiAsDriver(TuiHintManual)],
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { tuiHintManual: [{
                type: Input
            }] } });
