// @taiga-ui/core/esm2022/directives/hint/hint-describe.directive.mjs
import { __decorate } from "tslib";
import { DOCUMENT } from '@angular/common';
import { Directive, inject, Input, NgZone } from '@angular/core';
import { tuiIfMap, tuiTypedFromEvent, tuiZonefreeScheduler, tuiZoneOptimized, } from '@taiga-ui/cdk/observables';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { tuiIsNativeFocused } from '@taiga-ui/cdk/utils/focus';
import { tuiIsPresent, tuiPure } from '@taiga-ui/cdk/utils/miscellaneous';
import { tuiAsDriver, TuiDriver } from '@taiga-ui/core/classes';
import { BehaviorSubject, debounce, distinctUntilChanged, fromEvent, map, merge, of, skip, startWith, switchMap, timer, } from 'rxjs';
import * as i0 from "@angular/core";
class TuiHintDescribe extends TuiDriver {
    constructor() {
        super((subscriber) => this.stream$.subscribe(subscriber));
        this.doc = inject(DOCUMENT);
        this.el = tuiInjectElement();
        this.zone = inject(NgZone);
        this.id$ = new BehaviorSubject('');
        this.stream$ = this.id$.pipe(distinctUntilChanged(), tuiIfMap(() => fromEvent(this.doc, 'keydown', { capture: true }), tuiIsPresent), switchMap(() => this.focused
            ? of(false)
            : merge(tuiTypedFromEvent(this.doc, 'keyup'), tuiTypedFromEvent(this.element, 'blur')).pipe(map(() => this.focused))), debounce((visible) => visible ? timer(1000, tuiZonefreeScheduler(this.zone)) : of(null)), startWith(false), distinctUntilChanged(), skip(1), tuiZoneOptimized());
        this.type = 'hint';
    }
    set tuiHintDescribe(id) {
        this.id$.next(id || '');
    }
    get element() {
        return this.doc.getElementById(this.id$.value || '') || this.el;
    }
    get focused() {
        return tuiIsNativeFocused(this.element);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiHintDescribe, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiHintDescribe, isStandalone: true, selector: "[tuiHintDescribe]", inputs: { tuiHintDescribe: "tuiHintDescribe" }, providers: [tuiAsDriver(TuiHintDescribe)], usesInheritance: true, ngImport: i0 }); }
}
__decorate([
    tuiPure
], TuiHintDescribe.prototype, "element", null);
export { TuiHintDescribe };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiHintDescribe, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiHintDescribe]',
                    providers: [tuiAsDriver(TuiHintDescribe)],
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { tuiHintDescribe: [{
                type: Input
            }], element: [] } });
