// @taiga-ui/legacy/esm2022/directives/value-accessor/value-accessor.directive.mjs
import { Directive } from '@angular/core';
import { TUI_VALUE_ACCESSOR_PROVIDER } from './value-accessor.provider';
import * as i0 from "@angular/core";
class TuiValueAccessorDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiValueAccessorDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiValueAccessorDirective, selector: "[tuiValueAccessor]", providers: [TUI_VALUE_ACCESSOR_PROVIDER], ngImport: i0 }); }
}
export { TuiValueAccessorDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiValueAccessorDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: false,
                    selector: '[tuiValueAccessor]',
                    providers: [TUI_VALUE_ACCESSOR_PROVIDER],
                }]
        }] });
