// @taiga-ui/kit/fesm2022/taiga-ui-kit-components-status.mjs
import * as i0 from '@angular/core';
import { Component, ViewEncapsulation, ChangeDetectionStrategy, Directive } from '@angular/core';
import { tuiWithStyles } from '@taiga-ui/cdk/utils/miscellaneous';

class TuiStatusStyles {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiStatusStyles, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiStatusStyles, isStandalone: true, selector: "ng-component", host: { classAttribute: "tui-status" }, ngImport: i0, template: '', isInline: true, styles: ["[tuiStatus]{display:inline-flex;align-items:center;gap:.5rem}[tuiStatus]:before{content:\"\";display:var(--t-status, none);inline-size:.5rem;block-size:.5rem;border-radius:100%;background:var(--t-status)}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiStatusStyles, decorators: [{
            type: Component,
            args: [{ standalone: true, template: '', encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, host: {
                        class: 'tui-status',
                    }, styles: ["[tuiStatus]{display:inline-flex;align-items:center;gap:.5rem}[tuiStatus]:before{content:\"\";display:var(--t-status, none);inline-size:.5rem;block-size:.5rem;border-radius:100%;background:var(--t-status)}\n"] }]
        }] });
class TuiStatus {
    constructor() {
        this.nothing = tuiWithStyles(TuiStatusStyles);
        this.tuiStatus = '';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiStatus, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiStatus, isStandalone: true, selector: "[tuiStatus]", inputs: { tuiStatus: "tuiStatus" }, host: { attributes: { "tuiStatus": "" }, properties: { "style.--t-status": "tuiStatus || null" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiStatus, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiStatus]',
                    inputs: ['tuiStatus'],
                    host: {
                        tuiStatus: '',
                        '[style.--t-status]': 'tuiStatus || null',
                    },
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiStatus };
