// @taiga-ui/cdk/esm2022/directives/pan/pan.directive.mjs
import { Directive, inject, Output } from '@angular/core';
import { TuiPanService } from './pan.service';
import * as i0 from "@angular/core";
class TuiPan {
    constructor() {
        this.tuiPan = inject(TuiPanService);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPan, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiPan, isStandalone: true, selector: "[tuiPan]", outputs: { tuiPan: "tuiPan" }, providers: [TuiPanService], ngImport: i0 }); }
}
export { TuiPan };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPan, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiPan]',
                    providers: [TuiPanService],
                }]
        }], propDecorators: { tuiPan: [{
                type: Output
            }] } });
