// @taiga-ui/kit/esm2022/components/carousel/carousel-autoscroll.directive.mjs
import { Directive, inject, Output } from '@angular/core';
import { TuiCarouselDirective } from './carousel.directive';
import * as i0 from "@angular/core";
class TuiCarouselAutoscroll {
    constructor() {
        this.tuiCarouselAutoscroll = inject(TuiCarouselDirective);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiCarouselAutoscroll, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiCarouselAutoscroll, isStandalone: true, selector: "[tuiCarouselAutoscroll]", outputs: { tuiCarouselAutoscroll: "tuiCarouselAutoscroll" }, ngImport: i0 }); }
}
export { TuiCarouselAutoscroll };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiCarouselAutoscroll, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiCarouselAutoscroll]',
                }]
        }], propDecorators: { tuiCarouselAutoscroll: [{
                type: Output
            }] } });
