// @taiga-ui/kit/esm2022/components/carousel/carousel-buttons.directive.mjs
import { Directive } from '@angular/core';
import { tuiButtonOptionsProvider } from '@taiga-ui/core/components/button';
import * as i0 from "@angular/core";
class TuiCarouselButtons {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiCarouselButtons, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiCarouselButtons, isStandalone: true, selector: "[tuiCarouselButtons]", providers: [
            tuiButtonOptionsProvider({
                appearance: 'secondary',
                size: 'm',
            }),
        ], ngImport: i0 }); }
}
export { TuiCarouselButtons };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiCarouselButtons, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiCarouselButtons]',
                    providers: [
                        tuiButtonOptionsProvider({
                            appearance: 'secondary',
                            size: 'm',
                        }),
                    ],
                }]
        }] });
