// @taiga-ui/polymorpheus/esm2022/directives/template.mjs
import { ChangeDetectorRef, Directive, inject, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * ng-template wrapper directive also stores {@link ChangeDetectorRef} to properly handle change detection.
 */
class PolymorpheusTemplate {
    constructor(template = inject((TemplateRef), { self: true }), cdr = inject(ChangeDetectorRef)) {
        this.template = template;
        this.cdr = cdr;
        this.polymorpheus = '';
    }
    static ngTemplateContextGuard(_dir, _ctx) {
        return true;
    }
    check() {
        this.cdr.markForCheck();
    }
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: PolymorpheusTemplate, deps: [{ token: i0.TemplateRef }, { token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    /** @nocollapse */ static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: PolymorpheusTemplate, isStandalone: true, selector: "ng-template[polymorpheus]", inputs: { polymorpheus: "polymorpheus" }, exportAs: ["polymorpheus"], ngImport: i0 }); }
}
export { PolymorpheusTemplate };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: PolymorpheusTemplate, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'ng-template[polymorpheus]',
                    inputs: ['polymorpheus'],
                    exportAs: 'polymorpheus',
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }, { type: i0.ChangeDetectorRef }]; } });
