// @taiga-ui/kit/esm2022/components/input-phone-international/get-country-calling-code.pipe.mjs
import { Pipe } from '@angular/core';
import { CHAR_PLUS } from '@taiga-ui/cdk/constants';
import { getCountryCallingCode } from 'libphonenumber-js/core';
import * as i0 from "@angular/core";
class TuiGetCountryCallingCodePipe {
    transform(countryIsoCode, metadata) {
        return metadata
            ? CHAR_PLUS + getCountryCallingCode(countryIsoCode, metadata)
            : '';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiGetCountryCallingCodePipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiGetCountryCallingCodePipe, isStandalone: true, name: "tuiGetCountryCallingCode" }); }
}
export { TuiGetCountryCallingCodePipe };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiGetCountryCallingCodePipe, decorators: [{
            type: Pipe,
            args: [{
                    standalone: true,
                    name: 'tuiGetCountryCallingCode',
                }]
        }] });
