// @taiga-ui/legacy/esm2022/components/select-option/select-option.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiIcon } from '@taiga-ui/core/components/icon';
import { TuiScrollIntoView } from '@taiga-ui/core/components/scrollbar';
import { TuiSelectOptionComponent } from './select-option.component';
import * as i0 from "@angular/core";
class TuiSelectOptionModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSelectOptionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiSelectOptionModule, declarations: [TuiSelectOptionComponent], imports: [CommonModule, TuiScrollIntoView, TuiIcon], exports: [TuiSelectOptionComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSelectOptionModule, imports: [CommonModule, TuiIcon] }); }
}
export { TuiSelectOptionModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSelectOptionModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiScrollIntoView, TuiIcon],
                    declarations: [TuiSelectOptionComponent],
                    exports: [TuiSelectOptionComponent],
                }]
        }] });
