// @taiga-ui/legacy/esm2022/components/combo-box/combo-box.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiDropdown } from '@taiga-ui/core/directives/dropdown';
import { TuiArrowComponent } from '@taiga-ui/legacy/components/arrow';
import { TuiPrimitiveTextfieldModule, TuiTextfieldComponent, } from '@taiga-ui/legacy/components/primitive-textfield';
import { TuiSelectOptionModule } from '@taiga-ui/legacy/components/select-option';
import { TuiLegacyDropdownOpenMonitorDirective, TuiTextfieldControllerModule, TuiWrapperModule, } from '@taiga-ui/legacy/directives';
import { PolymorpheusOutlet, PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import { TuiComboBoxComponent } from './combo-box.component';
import { TuiComboBoxDirective } from './combo-box.directive';
import { TuiComboBoxStrictDirective } from './combo-box-strict.directive';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/dropdown";
class TuiComboBoxModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiComboBoxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiComboBoxModule, declarations: [TuiComboBoxComponent,
            TuiComboBoxStrictDirective,
            TuiComboBoxDirective], imports: [CommonModule,
            PolymorpheusOutlet,
            PolymorpheusTemplate,
            TuiPrimitiveTextfieldModule,
            TuiSelectOptionModule,
            TuiArrowComponent,
            TuiWrapperModule,
            TuiTextfieldControllerModule,
            TuiLegacyDropdownOpenMonitorDirective, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection], exports: [TuiComboBoxComponent,
            TuiComboBoxStrictDirective,
            TuiComboBoxDirective,
            TuiTextfieldComponent, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiComboBoxModule, imports: [CommonModule,
            TuiPrimitiveTextfieldModule,
            TuiSelectOptionModule,
            TuiArrowComponent,
            TuiWrapperModule,
            TuiTextfieldControllerModule, i1.TuiDropdownComponent] }); }
}
export { TuiComboBoxModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiComboBoxModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PolymorpheusOutlet,
                        PolymorpheusTemplate,
                        TuiPrimitiveTextfieldModule,
                        TuiSelectOptionModule,
                        TuiArrowComponent,
                        TuiWrapperModule,
                        TuiTextfieldControllerModule,
                        TuiLegacyDropdownOpenMonitorDirective,
                        ...TuiDropdown,
                    ],
                    declarations: [
                        TuiComboBoxComponent,
                        TuiComboBoxStrictDirective,
                        TuiComboBoxDirective,
                    ],
                    exports: [
                        TuiComboBoxComponent,
                        TuiComboBoxStrictDirective,
                        TuiComboBoxDirective,
                        TuiTextfieldComponent,
                        ...TuiDropdown,
                    ],
                }]
        }] });
