// @taiga-ui/kit/esm2022/components/action-bar/action-bar.directive.mjs
import { Directive } from '@angular/core';
import { TuiDropdownPortal } from '@taiga-ui/core/directives/dropdown';
import * as i0 from "@angular/core";
class TuiActionBarDirective extends TuiDropdownPortal {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiActionBarDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiActionBarDirective, isStandalone: true, selector: "ng-template[tuiActionBar]", inputs: { tuiDropdown: ["tuiActionBar", "tuiDropdown"] }, usesInheritance: true, ngImport: i0 }); }
}
export { TuiActionBarDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiActionBarDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'ng-template[tuiActionBar]',
                    inputs: ['tuiDropdown: tuiActionBar'],
                }]
        }] });
