// @taiga-ui/cdk/schematics/ng-update/v4/steps/update-packages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updatePackages = exports.TUI_EDITOR_VERSION = exports.TUI_EVENT_PLUGINS_VERSION = exports.TUI_DOMPURIFY_VERSION = exports.TUI_POLYMORPHEUS_VERSION = void 0;
const ng_morph_1 = require("ng-morph");
const version_1 = require("../../../../constants/version");
const constants_1 = require("../../../constants");
const steps_1 = require("../../steps");
exports.TUI_POLYMORPHEUS_VERSION = '^4.6.4';
exports.TUI_DOMPURIFY_VERSION = '^4.1.2';
exports.TUI_EVENT_PLUGINS_VERSION = '^4.0.1';
exports.TUI_EDITOR_VERSION = '^2.5.0';
function updatePackages({ tree }, _) {
    const packagesToRemove = ['@taiga-ui/addon-tablebars', '@taiga-ui/addon-preview'];
    packagesToRemove.forEach((pkg) => {
        (0, ng_morph_1.removePackageJsonDependency)(tree, pkg);
    });
    (0, steps_1.replacePackageName)('@tinkoff/ng-polymorpheus', {
        name: '@taiga-ui/polymorpheus',
        version: exports.TUI_POLYMORPHEUS_VERSION,
    }, tree);
    (0, steps_1.replacePackageName)('@tinkoff/ng-dompurify', {
        name: '@taiga-ui/dompurify',
        version: exports.TUI_DOMPURIFY_VERSION,
    }, tree);
    (0, steps_1.replacePackageName)('@tinkoff/ng-event-plugins', {
        name: '@taiga-ui/event-plugins',
        version: exports.TUI_EVENT_PLUGINS_VERSION,
    }, tree);
    (0, steps_1.replacePackageName)('@tinkoff/tui-editor', {
        name: '@taiga-ui/editor',
        version: exports.TUI_EDITOR_VERSION,
    }, tree);
    const cdk = (0, ng_morph_1.getPackageJsonDependency)(tree, '@taiga-ui/cdk');
    if (!(0, ng_morph_1.getPackageJsonDependency)(tree, '@taiga-ui/event-plugins')) {
        (0, ng_morph_1.addPackageJsonDependency)(tree, {
            name: '@taiga-ui/event-plugins',
            version: exports.TUI_EVENT_PLUGINS_VERSION,
            type: cdk === null || cdk === void 0 ? void 0 : cdk.type,
        });
    }
    if ((0, ng_morph_1.getImports)(constants_1.ALL_TS_FILES, { moduleSpecifier: '@taiga-ui/layout' }).length) {
        (0, ng_morph_1.addPackageJsonDependency)(tree, {
            name: '@taiga-ui/layout',
            version: version_1.TUI_VERSION,
            type: cdk === null || cdk === void 0 ? void 0 : cdk.type,
        });
    }
    if ((0, ng_morph_1.getImports)(constants_1.ALL_TS_FILES, { moduleSpecifier: '@taiga-ui/legacy' }).length) {
        (0, ng_morph_1.addPackageJsonDependency)(tree, {
            name: '@taiga-ui/legacy',
            version: version_1.TUI_VERSION,
            type: cdk === null || cdk === void 0 ? void 0 : cdk.type,
        });
    }
}
exports.updatePackages = updatePackages;
