// @taiga-ui/legacy/fesm2022/taiga-ui-legacy-components-island.mjs
import * as i0 from '@angular/core';
import { Directive, Input } from '@angular/core';

/**
 * @deprecated: drop in v5.0 use {@link TuiCardLarge}
 * https://taiga-ui.dev/layout/card-large
 */
class TuiIslandDirective {
    constructor() {
        this.size = 'm';
        this.textAlign = 'left';
        this.hoverable = false;
        this.transparent = false;
    }
    get sizeS() {
        return this.size === 's';
    }
    get sizeM() {
        return this.size === 'm';
    }
    get sizeL() {
        return this.size === 'l';
    }
    get textAlignLeft() {
        return this.textAlign === 'left';
    }
    get textAlignCenter() {
        return this.textAlign === 'center';
    }
    get textAlignRight() {
        return this.textAlign === 'right';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiIslandDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiIslandDirective, isStandalone: true, selector: "tui-island, a[tuiIsland]", inputs: { size: "size", textAlign: "textAlign", hoverable: "hoverable", transparent: "transparent" }, host: { properties: { "class.tui-island_hoverable": "hoverable", "class.tui-island_transparent": "transparent", "class.tui-island_size_s": "sizeS", "class.tui-island_size_m": "sizeM", "class.tui-island_size_l": "sizeL", "class.tui-island_text-align_left": "textAlignLeft", "class.tui-island_text-align_center": "textAlignCenter", "class.tui-island_text-align_right": "textAlignRight" }, classAttribute: "tui-island" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiIslandDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'tui-island, a[tuiIsland]',
                    host: {
                        class: 'tui-island',
                        '[class.tui-island_hoverable]': 'hoverable',
                        '[class.tui-island_transparent]': 'transparent',
                        '[class.tui-island_size_s]': 'sizeS',
                        '[class.tui-island_size_m]': 'sizeM',
                        '[class.tui-island_size_l]': 'sizeL',
                        '[class.tui-island_text-align_left]': 'textAlignLeft',
                        '[class.tui-island_text-align_center]': 'textAlignCenter',
                        '[class.tui-island_text-align_right]': 'textAlignRight',
                    },
                }]
        }], propDecorators: { size: [{
                type: Input
            }], textAlign: [{
                type: Input
            }], hoverable: [{
                type: Input
            }], transparent: [{
                type: Input
            }] } });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiIslandDirective };
