// @taiga-ui/legacy/fesm2022/taiga-ui-legacy-components-multi-select-option.mjs
import * as i0 from '@angular/core';
import { Component, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { tuiIsPresent } from '@taiga-ui/cdk/utils/miscellaneous';
import { TuiSelectOptionComponent } from '@taiga-ui/legacy/components/select-option';
import * as i1 from '@angular/common';
import { CommonModule } from '@angular/common';
import * as i2 from '@taiga-ui/kit/components/checkbox';
import { TuiCheckbox } from '@taiga-ui/kit/components/checkbox';

class TuiMultiSelectOptionComponent extends TuiSelectOptionComponent {
    get size() {
        return this.dataList?.size === 'l' ? 'm' : 's';
    }
    get selected() {
        const { value } = this.option;
        return (tuiIsPresent(value) &&
            tuiIsPresent(this.value) &&
            this.value.some((item) => this.matcher(item, value)));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectOptionComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiMultiSelectOptionComponent, selector: "tui-multi-select-option", usesInheritance: true, ngImport: i0, template: "<input\n    tuiCheckbox\n    type=\"checkbox\"\n    class=\"t-checkbox\"\n    [checked]=\"selected$ | async\"\n    [size]=\"size\"\n/>\n<ng-container [ngTemplateOutlet]=\"context.$implicit\" />\n", styles: ["@keyframes retrigger{0%{left:1px}to{left:2px}}:host{display:flex;flex:1;align-items:center;animation:retrigger 1s}.t-checkbox{margin-right:.75rem}\n"], dependencies: [{ kind: "directive", type: i1.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet", "ngTemplateOutletInjector"] }, { kind: "component", type: i2.TuiCheckbox, selector: "input[type=\"checkbox\"][tuiCheckbox]", inputs: ["size"] }, { kind: "pipe", type: i1.AsyncPipe, name: "async" }], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectOptionComponent, decorators: [{
            type: Component,
            args: [{ standalone: false, selector: 'tui-multi-select-option', changeDetection: ChangeDetectionStrategy.OnPush, template: "<input\n    tuiCheckbox\n    type=\"checkbox\"\n    class=\"t-checkbox\"\n    [checked]=\"selected$ | async\"\n    [size]=\"size\"\n/>\n<ng-container [ngTemplateOutlet]=\"context.$implicit\" />\n", styles: ["@keyframes retrigger{0%{left:1px}to{left:2px}}:host{display:flex;flex:1;align-items:center;animation:retrigger 1s}.t-checkbox{margin-right:.75rem}\n"] }]
        }] });

class TuiMultiSelectOptionModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectOptionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectOptionModule, declarations: [TuiMultiSelectOptionComponent], imports: [CommonModule, TuiCheckbox], exports: [TuiMultiSelectOptionComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectOptionModule, imports: [CommonModule, TuiCheckbox] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectOptionModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiCheckbox],
                    declarations: [TuiMultiSelectOptionComponent],
                    exports: [TuiMultiSelectOptionComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiMultiSelectOptionComponent, TuiMultiSelectOptionModule };
