// @taiga-ui/legacy/esm2022/components/color-selector/color-edit/color-edit.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { MaskitoDirective } from '@maskito/angular';
import { TuiButton } from '@taiga-ui/core/components/button';
import { TuiDataList } from '@taiga-ui/core/components/data-list';
import { TuiDropdown } from '@taiga-ui/core/directives/dropdown';
import { TuiInputNumberModule } from '@taiga-ui/legacy/components/input-number';
import { TuiPrimitiveTextfieldModule } from '@taiga-ui/legacy/components/primitive-textfield';
import { TuiSelectModule } from '@taiga-ui/legacy/components/select';
import { TuiTextfieldControllerModule } from '@taiga-ui/legacy/directives/textfield-controller';
import { TuiColorEditComponent } from './color-edit.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/dropdown";
import * as i2 from "@taiga-ui/core/components/data-list";
class TuiColorEditModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiColorEditModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiColorEditModule, declarations: [TuiColorEditComponent], imports: [CommonModule,
            FormsModule,
            MaskitoDirective,
            TuiPrimitiveTextfieldModule,
            TuiTextfieldControllerModule,
            TuiInputNumberModule,
            TuiSelectModule, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection, i2.TuiDataListComponent, i2.TuiDataListDirective, i2.TuiOption, i2.TuiOptGroup, TuiButton], exports: [TuiColorEditComponent, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection, i2.TuiDataListComponent, i2.TuiDataListDirective, i2.TuiOption, i2.TuiOptGroup] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiColorEditModule, imports: [CommonModule,
            FormsModule,
            TuiPrimitiveTextfieldModule,
            TuiTextfieldControllerModule,
            TuiInputNumberModule,
            TuiSelectModule, i1.TuiDropdownComponent, i2.TuiDataListComponent, i2.TuiOption] }); }
}
export { TuiColorEditModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiColorEditModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        MaskitoDirective,
                        TuiPrimitiveTextfieldModule,
                        TuiTextfieldControllerModule,
                        TuiInputNumberModule,
                        TuiSelectModule,
                        ...TuiDropdown,
                        ...TuiDataList,
                        TuiButton,
                    ],
                    declarations: [TuiColorEditComponent],
                    exports: [TuiColorEditComponent, ...TuiDropdown, ...TuiDataList],
                }]
        }] });
