// @taiga-ui/kit/esm2022/components/files/input-files/input-files-validator.directive.mjs
import { Directive, inject, Input } from '@angular/core';
import { Validators } from '@angular/forms';
import { TuiValidator } from '@taiga-ui/cdk/directives/validator';
import { tuiCreateFileFormatValidator, tuiCreateFileSizeValidator, } from '../files.validators';
import { TUI_INPUT_FILES_OPTIONS } from './input-files.options';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk/directives/validator";
class TuiInputFilesValidator {
    constructor() {
        this.options = inject(TUI_INPUT_FILES_OPTIONS);
        this.validator = inject(TuiValidator);
        this.accept = this.options.accept;
        this.maxFileSize = this.options.maxFileSize;
    }
    ngOnChanges() {
        this.validate();
    }
    ngOnInit() {
        this.validate();
    }
    validate() {
        this.validator.tuiValidator =
            Validators.compose([
                tuiCreateFileFormatValidator(this.accept),
                tuiCreateFileSizeValidator(this.maxFileSize),
            ]) || Validators.nullValidator;
        this.validator.ngOnChanges();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputFilesValidator, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiInputFilesValidator, isStandalone: true, inputs: { accept: "accept", maxFileSize: "maxFileSize" }, host: { properties: { "accept": "accept" } }, exportAs: ["tuiInputFilesValidator"], usesOnChanges: true, hostDirectives: [{ directive: i1.TuiValidator }], ngImport: i0 }); }
}
export { TuiInputFilesValidator };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputFilesValidator, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    exportAs: 'tuiInputFilesValidator',
                    hostDirectives: [TuiValidator],
                    host: {
                        '[accept]': 'accept',
                    },
                }]
        }], propDecorators: { accept: [{
                type: Input
            }], maxFileSize: [{
                type: Input
            }] } });
