// @taiga-ui/core/esm2022/components/textfield/textfield.options.mjs
import { Directive, inject, Input, Optional, signal, SkipSelf } from '@angular/core';
import { tuiCreateToken, tuiProvide } from '@taiga-ui/cdk/utils/miscellaneous';
import * as i0 from "@angular/core";
const DEFAULT = {
    appearance: 'textfield',
    size: 'l',
    cleaner: true,
};
export const TUI_TEXTFIELD_OPTIONS = tuiCreateToken({
    appearance: signal(DEFAULT.appearance),
    size: signal(DEFAULT.size),
    cleaner: signal(DEFAULT.cleaner),
});
export function tuiTextfieldOptionsProvider(options) {
    return {
        provide: TUI_TEXTFIELD_OPTIONS,
        deps: [[new Optional(), new SkipSelf(), TUI_TEXTFIELD_OPTIONS]],
        useFactory: (parent) => ({
            appearance: signal(parent?.appearance() ?? DEFAULT.appearance),
            size: signal(parent?.size() ?? DEFAULT.size),
            cleaner: signal(parent?.cleaner() ?? DEFAULT.cleaner),
            ...options,
        }),
    };
}
class TuiTextfieldOptionsDirective {
    constructor() {
        this.options = inject(TUI_TEXTFIELD_OPTIONS, { skipSelf: true });
        // TODO: refactor to signal inputs after Angular update
        this.appearance = signal(this.options.appearance());
        this.size = signal(this.options.size());
        this.cleaner = signal(this.options.cleaner());
    }
    set tuiTextfieldAppearance(appearance) {
        this.appearance.set(appearance);
    }
    set tuiTextfieldSize(size) {
        this.size.set(size);
    }
    set tuiTextfieldCleaner(enabled) {
        this.cleaner.set(enabled);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextfieldOptionsDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiTextfieldOptionsDirective, isStandalone: true, selector: "[tuiTextfieldAppearance],[tuiTextfieldSize],[tuiTextfieldCleaner]", inputs: { tuiTextfieldAppearance: "tuiTextfieldAppearance", tuiTextfieldSize: "tuiTextfieldSize", tuiTextfieldCleaner: "tuiTextfieldCleaner" }, providers: [tuiProvide(TUI_TEXTFIELD_OPTIONS, TuiTextfieldOptionsDirective)], ngImport: i0 }); }
}
export { TuiTextfieldOptionsDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextfieldOptionsDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiTextfieldAppearance],[tuiTextfieldSize],[tuiTextfieldCleaner]',
                    providers: [tuiProvide(TUI_TEXTFIELD_OPTIONS, TuiTextfieldOptionsDirective)],
                }]
        }], propDecorators: { tuiTextfieldAppearance: [{
                type: Input
            }], tuiTextfieldSize: [{
                type: Input
            }], tuiTextfieldCleaner: [{
                type: Input
            }] } });
