// @taiga-ui/legacy/esm2022/components/input-password/input-password.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiIcon } from '@taiga-ui/core/components/icon';
import { TuiAppearance } from '@taiga-ui/core/directives/appearance';
import { TuiHint } from '@taiga-ui/core/directives/hint';
import { TuiPrimitiveTextfieldModule, TuiTextfieldComponent, } from '@taiga-ui/legacy/components/primitive-textfield';
import { TuiTextfieldControllerModule } from '@taiga-ui/legacy/directives';
import { PolymorpheusOutlet, PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import { TuiInputPasswordComponent } from './input-password.component';
import { TuiInputPasswordDirective } from './input-password.directive';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/hint";
/**
 * @deprecated use {@link TuiPassword} with {@link TuiTextfield}
 */
class TuiInputPasswordModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputPasswordModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiInputPasswordModule, declarations: [TuiInputPasswordComponent, TuiInputPasswordDirective], imports: [i1.TuiHintComponent, i1.TuiHintDirective, i1.TuiHintOptionsDirective, i1.TuiHintUnstyled, i1.TuiHintDriver, i1.TuiHintPosition, i1.TuiHintHover, i1.TuiHintOverflow, i1.TuiHintDescribe, i1.TuiHintHost, i1.TuiHintManual, i1.TuiHintPointer, CommonModule,
            FormsModule,
            PolymorpheusOutlet,
            PolymorpheusTemplate,
            TuiPrimitiveTextfieldModule,
            TuiTextfieldControllerModule,
            TuiAppearance,
            TuiIcon], exports: [TuiInputPasswordComponent,
            TuiInputPasswordDirective,
            TuiTextfieldComponent, i1.TuiHintComponent, i1.TuiHintDirective, i1.TuiHintOptionsDirective, i1.TuiHintUnstyled, i1.TuiHintDriver, i1.TuiHintPosition, i1.TuiHintHover, i1.TuiHintOverflow, i1.TuiHintDescribe, i1.TuiHintHost, i1.TuiHintManual, i1.TuiHintPointer] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputPasswordModule, imports: [i1.TuiHintComponent, CommonModule,
            FormsModule,
            TuiPrimitiveTextfieldModule,
            TuiTextfieldControllerModule,
            TuiIcon] }); }
}
export { TuiInputPasswordModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputPasswordModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        ...TuiHint,
                        CommonModule,
                        FormsModule,
                        PolymorpheusOutlet,
                        PolymorpheusTemplate,
                        TuiPrimitiveTextfieldModule,
                        TuiTextfieldControllerModule,
                        TuiAppearance,
                        TuiIcon,
                    ],
                    declarations: [TuiInputPasswordComponent, TuiInputPasswordDirective],
                    exports: [
                        TuiInputPasswordComponent,
                        TuiInputPasswordDirective,
                        TuiTextfieldComponent,
                        ...TuiHint,
                    ],
                }]
        }] });
