// @taiga-ui/core/esm2022/components/data-list/opt-group.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
class TuiOptGroup {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiOptGroup, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiOptGroup, isStandalone: true, selector: "tui-opt-group", inputs: { label: "label" }, host: { attributes: { "role": "group" }, properties: { "attr.data-label": "label" } }, ngImport: i0 }); }
}
export { TuiOptGroup };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiOptGroup, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'tui-opt-group',
                    host: {
                        role: 'group',
                        '[attr.data-label]': 'label',
                    },
                }]
        }], propDecorators: { label: [{
                type: Input
            }] } });
