// @taiga-ui/core/esm2022/components/scrollbar/scroll-ref.directive.mjs
import { Directive, ElementRef } from '@angular/core';
import { tuiProvide } from '@taiga-ui/cdk/utils/miscellaneous';
import { TUI_SCROLL_REF } from '@taiga-ui/core/tokens';
import * as i0 from "@angular/core";
export const SCROLL_REF_SELECTOR = '[tuiScrollRef]';
class TuiScrollRef {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiScrollRef, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiScrollRef, isStandalone: true, selector: "[tuiScrollRef]", providers: [tuiProvide(TUI_SCROLL_REF, ElementRef)], ngImport: i0 }); }
}
export { TuiScrollRef };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiScrollRef, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: SCROLL_REF_SELECTOR,
                    providers: [tuiProvide(TUI_SCROLL_REF, ElementRef)],
                }]
        }] });
