// @taiga-ui/core/esm2022/components/scrollbar/scrollable.directive.mjs
import { Directive } from '@angular/core';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { TUI_SCROLLABLE } from './scrollbar.component';
import * as i0 from "@angular/core";
class TuiScrollable {
    constructor() {
        this.el = tuiInjectElement();
    }
    ngOnInit() {
        this.el.dispatchEvent(new CustomEvent(TUI_SCROLLABLE, {
            bubbles: true,
            detail: this.el,
        }));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiScrollable, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiScrollable, isStandalone: true, selector: "[tuiScrollable]", ngImport: i0 }); }
}
export { TuiScrollable };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiScrollable, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiScrollable]',
                }]
        }] });
