// @taiga-ui/core/esm2022/components/scrollbar/scroll-into-view.directive.mjs
import { DestroyRef, Directive, inject, Input } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { timer } from 'rxjs';
import { TUI_SCROLL_INTO_VIEW } from './scrollbar.component';
import * as i0 from "@angular/core";
/**
 * Directive scrolls element into view inside tui-scrollbar
 */
class TuiScrollIntoView {
    constructor() {
        this.el = tuiInjectElement();
        this.destroyRef = inject(DestroyRef);
    }
    set tuiScrollIntoView(scroll) {
        if (!scroll) {
            return;
        }
        // Timeout is necessary in order to give element render cycle to get into its final spot
        // (for example if it is inside dropdown box which has to be positioned first)
        timer(0)
            .pipe(takeUntilDestroyed(this.destroyRef))
            .subscribe(() => {
            this.el.dispatchEvent(new CustomEvent(TUI_SCROLL_INTO_VIEW, {
                bubbles: true,
                detail: this.el,
            }));
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiScrollIntoView, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiScrollIntoView, isStandalone: true, selector: "[tuiScrollIntoView]", inputs: { tuiScrollIntoView: "tuiScrollIntoView" }, ngImport: i0 }); }
}
export { TuiScrollIntoView };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiScrollIntoView, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiScrollIntoView]',
                }]
        }], propDecorators: { tuiScrollIntoView: [{
                type: Input
            }] } });
