// @taiga-ui/cdk/esm2022/directives/native-validator/native-validator.directive.mjs
import { DestroyRef, Directive, inject, Input, NgZone } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { tuiTakeUntilDestroyed, tuiZonefree } from '@taiga-ui/cdk/observables';
import { tuiInjectElement, tuiProvide } from '@taiga-ui/cdk/utils';
import { timer } from 'rxjs';
import * as i0 from "@angular/core";
class TuiNativeValidator {
    constructor() {
        this.destroyRef = inject(DestroyRef);
        this.zone = inject(NgZone);
        this.host = tuiInjectElement();
        this.tuiNativeValidator = 'Invalid';
    }
    validate(control) {
        this.control = control;
        timer(0)
            .pipe(tuiZonefree(this.zone), tuiTakeUntilDestroyed(this.destroyRef))
            .subscribe(() => this.handleValidation());
        return null;
    }
    handleValidation() {
        this.el.setCustomValidity?.(this.control?.touched && this.control?.invalid ? this.tuiNativeValidator : '');
    }
    get el() {
        return this.host.querySelector('input,textarea,select') || this.host;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiNativeValidator, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiNativeValidator, isStandalone: true, selector: "[tuiNativeValidator]", inputs: { tuiNativeValidator: "tuiNativeValidator" }, host: { listeners: { "blur": "handleValidation()" } }, providers: [tuiProvide(NG_VALIDATORS, TuiNativeValidator, true)], ngImport: i0 }); }
}
export { TuiNativeValidator };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiNativeValidator, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiNativeValidator]',
                    providers: [tuiProvide(NG_VALIDATORS, TuiNativeValidator, true)],
                    host: {
                        '(blur)': 'handleValidation()',
                    },
                }]
        }], propDecorators: { tuiNativeValidator: [{
                type: Input
            }] } });
