// @taiga-ui/legacy/esm2022/components/input-range/input-range.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiActiveZone } from '@taiga-ui/cdk/directives/active-zone';
import { TuiNumberFormat } from '@taiga-ui/core/directives/number-format';
import { TuiRange } from '@taiga-ui/kit/components/range';
import { TuiInputNumberModule } from '@taiga-ui/legacy/components/input-number';
import { TuiTextfieldControllerModule, TuiWrapperModule, } from '@taiga-ui/legacy/directives';
import { PolymorpheusOutlet, PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import { TuiInputRangeComponent } from './input-range.component';
import * as i0 from "@angular/core";
class TuiInputRangeModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputRangeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiInputRangeModule, declarations: [TuiInputRangeComponent], imports: [CommonModule,
            FormsModule,
            PolymorpheusOutlet,
            PolymorpheusTemplate,
            TuiActiveZone,
            TuiInputNumberModule,
            TuiRange,
            TuiWrapperModule,
            TuiTextfieldControllerModule,
            TuiNumberFormat], exports: [TuiInputRangeComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputRangeModule, imports: [CommonModule,
            FormsModule,
            TuiInputNumberModule,
            TuiRange,
            TuiWrapperModule,
            TuiTextfieldControllerModule] }); }
}
export { TuiInputRangeModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputRangeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        PolymorpheusOutlet,
                        PolymorpheusTemplate,
                        TuiActiveZone,
                        TuiInputNumberModule,
                        TuiRange,
                        TuiWrapperModule,
                        TuiTextfieldControllerModule,
                        TuiNumberFormat,
                    ],
                    declarations: [TuiInputRangeComponent],
                    exports: [TuiInputRangeComponent],
                }]
        }] });
