// @taiga-ui/cdk/esm2022/directives/zoom/zoom.directive.mjs
import { Directive, inject } from '@angular/core';
import { TuiZoomService } from './zoom.service';
import * as i0 from "@angular/core";
class TuiZoom {
    constructor() {
        this.tuiZoom = inject(TuiZoomService);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiZoom, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiZoom, isStandalone: true, selector: "[tuiZoom]", outputs: { tuiZoom: "tuiZoom" }, host: { properties: { "style.touch-action": "\"none\"" } }, providers: [TuiZoomService], ngImport: i0 }); }
}
export { TuiZoom };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiZoom, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiZoom]',
                    outputs: ['tuiZoom'],
                    providers: [TuiZoomService],
                    host: {
                        '[style.touch-action]': '"none"',
                    },
                }]
        }] });
