// @taiga-ui/cdk/esm2022/directives/auto-focus/autofocus.options.mjs
import { ElementRef, InjectionToken, NgZone, Renderer2 } from '@angular/core';
import { WA_ANIMATION_FRAME, WA_WINDOW } from '@ng-web-apis/common';
import { TUI_IS_IOS } from '@taiga-ui/cdk/tokens';
import { tuiCreateToken, tuiProvideOptions } from '@taiga-ui/cdk/utils/miscellaneous';
import { TuiDefaultAutofocusHandler } from './handlers/default.handler';
import { TuiIosAutofocusHandler } from './handlers/ios.handler';
export const TUI_AUTOFOCUS_DEFAULT_OPTIONS = {
    delay: NaN, // NaN = no delay/sync
};
export const TUI_AUTOFOCUS_OPTIONS = tuiCreateToken(TUI_AUTOFOCUS_DEFAULT_OPTIONS);
export function tuiAutoFocusOptionsProvider(options) {
    return tuiProvideOptions(TUI_AUTOFOCUS_OPTIONS, options, TUI_AUTOFOCUS_DEFAULT_OPTIONS);
}
export const TUI_AUTOFOCUS_HANDLER = new InjectionToken('[TUI_AUTOFOCUS_HANDLER]');
export const TUI_AUTOFOCUS_PROVIDERS = [
    {
        provide: TUI_AUTOFOCUS_HANDLER,
        useFactory: (el, animationFrame$, renderer, zone, win, isIos) => isIos
            ? new TuiIosAutofocusHandler(el, renderer, zone, win)
            : new TuiDefaultAutofocusHandler(el, animationFrame$, zone),
        deps: [ElementRef, WA_ANIMATION_FRAME, Renderer2, NgZone, WA_WINDOW, TUI_IS_IOS],
    },
];
