// @taiga-ui/legacy/esm2022/classes/abstract-textfield-host.mjs
import { Directive, inject } from '@angular/core';
import { AbstractTuiControl } from './control';
import * as i0 from "@angular/core";
/**
 * @deprecated: drop in v5.0
 */
class AbstractTuiTextfieldHost {
    constructor() {
        this.host = inject(AbstractTuiControl, { optional: true });
    }
    get readOnly() {
        return this.host.readOnly;
    }
    get disabled() {
        return this.host.computedDisabled;
    }
    get invalid() {
        return this.host.computedInvalid;
    }
    get focusable() {
        return this.host.computedFocusable;
    }
    get inputMode() {
        return 'text';
    }
    get value() {
        return this.host.value?.toString() || '';
    }
    process(_input) { }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: AbstractTuiTextfieldHost, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: AbstractTuiTextfieldHost, ngImport: i0 }); }
}
export { AbstractTuiTextfieldHost };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: AbstractTuiTextfieldHost, decorators: [{
            type: Directive
        }] });
