// @taiga-ui/legacy/esm2022/classes/abstract-native-select.mjs
import { Directive, inject, Input, TemplateRef, ViewChild } from '@angular/core';
import { tuiInjectId } from '@taiga-ui/cdk/services';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { TuiDataListDirective } from '@taiga-ui/core/components/data-list';
import { TUI_ITEMS_HANDLERS } from '@taiga-ui/kit/tokens';
import { TUI_TEXTFIELD_HOST } from '@taiga-ui/legacy/tokens';
import { AbstractTuiControl } from './control';
import * as i0 from "@angular/core";
/**
 * @deprecated: drop in v5.0
 */
class AbstractTuiNativeSelect {
    constructor() {
        this.datalist = null;
        this.autoId = tuiInjectId();
        this.el = tuiInjectElement();
        this.host = inject(TUI_TEXTFIELD_HOST);
        this.control = inject(AbstractTuiControl);
        this.itemsHandlers = inject(TUI_ITEMS_HANDLERS);
        this.disabledItemHandler = null;
        this.placeholder = '';
    }
    get id() {
        return this.el.id || this.autoId;
    }
    get emptyOption() {
        return !!this.placeholder && !this.control.value;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: AbstractTuiNativeSelect, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: AbstractTuiNativeSelect, inputs: { disabledItemHandler: "disabledItemHandler", placeholder: "placeholder" }, host: { properties: { "id": "id" } }, viewQueries: [{ propertyName: "datalist", first: true, predicate: TuiDataListDirective, descendants: true, read: TemplateRef, static: true }], ngImport: i0 }); }
}
export { AbstractTuiNativeSelect };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: AbstractTuiNativeSelect, decorators: [{
            type: Directive,
            args: [{
                    standalone: false,
                    host: {
                        '[id]': 'id',
                    },
                }]
        }], propDecorators: { datalist: [{
                type: ViewChild,
                args: [TuiDataListDirective, { read: TemplateRef, static: true }]
            }], disabledItemHandler: [{
                type: Input
            }], placeholder: [{
                type: Input
            }] } });
