// @taiga-ui/cdk/esm2022/directives/popover/popover.directive.mjs
import { Directive, inject } from '@angular/core';
import { tuiIfMap } from '@taiga-ui/cdk/observables';
import { TuiPopoverService } from '@taiga-ui/cdk/services';
import { PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import { distinctUntilChanged, endWith, ignoreElements, share, Subject } from 'rxjs';
import * as i0 from "@angular/core";
class TuiPopoverDirective extends PolymorpheusTemplate {
    constructor() {
        super(...arguments);
        this.service = inject((TuiPopoverService));
        this.open$ = new Subject();
        this.options = {};
        this.open = false;
        this.openChange = this.open$.pipe(distinctUntilChanged(), tuiIfMap(() => this.service.open(this, this.options).pipe(ignoreElements(), endWith(false))), share());
    }
    ngOnChanges() {
        this.open$.next(this.open);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPopoverDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiPopoverDirective, usesInheritance: true, usesOnChanges: true, ngImport: i0 }); }
}
export { TuiPopoverDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPopoverDirective, decorators: [{
            type: Directive
        }] });
