// @taiga-ui/legacy/esm2022/components/textarea/textarea.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiIcon } from '@taiga-ui/core/components/icon';
import { TuiScrollbar } from '@taiga-ui/core/components/scrollbar';
import { TuiAppearance } from '@taiga-ui/core/directives/appearance';
import { TuiPrimitiveTextfieldModule, TuiTextfieldComponent, } from '@taiga-ui/legacy/components/primitive-textfield';
import { TuiTooltipModule } from '@taiga-ui/legacy/components/tooltip';
import { TuiWrapperModule } from '@taiga-ui/legacy/directives';
import { PolymorpheusOutlet, PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import { TuiTextareaComponent } from './textarea.component';
import { TuiTextareaDirective } from './textarea.directive';
import * as i0 from "@angular/core";
class TuiTextareaModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextareaModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiTextareaModule, declarations: [TuiTextareaComponent, TuiTextareaDirective], imports: [CommonModule,
            FormsModule,
            TuiScrollbar,
            TuiTooltipModule,
            TuiWrapperModule,
            TuiPrimitiveTextfieldModule,
            PolymorpheusOutlet,
            PolymorpheusTemplate,
            TuiIcon,
            TuiAppearance], exports: [TuiTextareaComponent,
            TuiTextareaDirective,
            TuiTextfieldComponent,
            PolymorpheusOutlet,
            PolymorpheusTemplate] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextareaModule, imports: [CommonModule,
            FormsModule,
            TuiScrollbar,
            TuiTooltipModule,
            TuiWrapperModule,
            TuiPrimitiveTextfieldModule,
            TuiIcon] }); }
}
export { TuiTextareaModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextareaModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        TuiScrollbar,
                        TuiTooltipModule,
                        TuiWrapperModule,
                        TuiPrimitiveTextfieldModule,
                        PolymorpheusOutlet,
                        PolymorpheusTemplate,
                        TuiIcon,
                        TuiAppearance,
                    ],
                    declarations: [TuiTextareaComponent, TuiTextareaDirective],
                    exports: [
                        TuiTextareaComponent,
                        TuiTextareaDirective,
                        TuiTextfieldComponent,
                        PolymorpheusOutlet,
                        PolymorpheusTemplate,
                    ],
                }]
        }] });
