// @taiga-ui/legacy/esm2022/components/input-date/input-date.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MaskitoDirective } from '@maskito/angular';
import { TuiLet } from '@taiga-ui/cdk/directives/let';
import { TuiCalendar } from '@taiga-ui/core/components/calendar';
import { TuiIcon } from '@taiga-ui/core/components/icon';
import { TuiLink } from '@taiga-ui/core/components/link';
import { TuiAppearance } from '@taiga-ui/core/directives/appearance';
import { TuiDropdown } from '@taiga-ui/core/directives/dropdown';
import { TuiPrimitiveTextfieldModule, TuiTextfieldComponent, } from '@taiga-ui/legacy/components/primitive-textfield';
import { TuiLegacyDropdownOpenMonitorDirective, TuiTextfieldControllerModule, TuiValueAccessorModule, } from '@taiga-ui/legacy/directives';
import { PolymorpheusOutlet, PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import { TuiInputDateComponent } from './input-date.component';
import { TuiInputDateDirective } from './input-date.directive';
import { TuiNativeDateDirective } from './native-date/native-date.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/dropdown";
class TuiInputDateModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateModule, declarations: [TuiInputDateComponent, TuiInputDateDirective, TuiNativeDateDirective], imports: [CommonModule,
            MaskitoDirective,
            PolymorpheusOutlet,
            PolymorpheusTemplate,
            TuiCalendar,
            TuiLink,
            TuiPrimitiveTextfieldModule,
            TuiValueAccessorModule,
            TuiLet,
            TuiTextfieldControllerModule,
            TuiLegacyDropdownOpenMonitorDirective,
            TuiAppearance,
            TuiIcon, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection], exports: [TuiInputDateComponent,
            TuiInputDateDirective,
            TuiTextfieldComponent, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateModule, imports: [CommonModule,
            TuiCalendar,
            TuiPrimitiveTextfieldModule,
            TuiValueAccessorModule,
            TuiTextfieldControllerModule,
            TuiIcon, i1.TuiDropdownComponent] }); }
}
export { TuiInputDateModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MaskitoDirective,
                        PolymorpheusOutlet,
                        PolymorpheusTemplate,
                        TuiCalendar,
                        TuiLink,
                        TuiPrimitiveTextfieldModule,
                        TuiValueAccessorModule,
                        TuiLet,
                        TuiTextfieldControllerModule,
                        TuiLegacyDropdownOpenMonitorDirective,
                        TuiAppearance,
                        TuiIcon,
                        ...TuiDropdown,
                    ],
                    declarations: [TuiInputDateComponent, TuiInputDateDirective, TuiNativeDateDirective],
                    exports: [
                        TuiInputDateComponent,
                        TuiInputDateDirective,
                        TuiTextfieldComponent,
                        ...TuiDropdown,
                    ],
                }]
        }] });
