// @taiga-ui/legacy/esm2022/components/input-year/input-year.module.mjs
import { NgModule } from '@angular/core';
import { MaskitoDirective } from '@maskito/angular';
import { TuiCalendarYear } from '@taiga-ui/core/components/calendar';
import { TuiScrollbar } from '@taiga-ui/core/components/scrollbar';
import { TuiDropdown, TuiDropdownOpen } from '@taiga-ui/core/directives/dropdown';
import { TuiPrimitiveTextfieldModule, TuiTextfieldComponent, } from '@taiga-ui/legacy/components/primitive-textfield';
import { TuiTextfieldControllerModule } from '@taiga-ui/legacy/directives';
import { TuiInputYearComponent } from './input-year.component';
import { TuiInputYearDirective } from './input-year.directive';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/dropdown";
class TuiInputYearModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputYearModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiInputYearModule, declarations: [TuiInputYearComponent, TuiInputYearDirective], imports: [MaskitoDirective,
            TuiDropdownOpen,
            TuiPrimitiveTextfieldModule,
            TuiScrollbar,
            TuiCalendarYear,
            TuiTextfieldControllerModule, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection], exports: [TuiInputYearComponent,
            TuiInputYearDirective,
            TuiTextfieldComponent, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputYearModule, imports: [TuiPrimitiveTextfieldModule,
            TuiScrollbar,
            TuiCalendarYear,
            TuiTextfieldControllerModule, i1.TuiDropdownComponent] }); }
}
export { TuiInputYearModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputYearModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        MaskitoDirective,
                        TuiDropdownOpen,
                        TuiPrimitiveTextfieldModule,
                        TuiScrollbar,
                        TuiCalendarYear,
                        TuiTextfieldControllerModule,
                        ...TuiDropdown,
                    ],
                    declarations: [TuiInputYearComponent, TuiInputYearDirective],
                    exports: [
                        TuiInputYearComponent,
                        TuiInputYearDirective,
                        TuiTextfieldComponent,
                        ...TuiDropdown,
                    ],
                }]
        }] });
