// @taiga-ui/legacy/esm2022/components/color-selector/color-selector.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiActiveZone } from '@taiga-ui/cdk/directives/active-zone';
import { TuiButton } from '@taiga-ui/core/components/button';
import { TuiDataList } from '@taiga-ui/core/components/data-list';
import { TuiIcon } from '@taiga-ui/core/components/icon';
import { TuiGroup } from '@taiga-ui/core/directives/group';
import { TuiHint } from '@taiga-ui/core/directives/hint';
import { TuiChevron } from '@taiga-ui/kit/directives/chevron';
import { TuiSvgComponent } from '@taiga-ui/legacy/components/svg';
import { TuiColorEditModule } from './color-edit/color-edit.module';
import { TuiColorPickerModule } from './color-picker/color-picker.module';
import { TuiColorSelectorComponent } from './color-selector.component';
import { TuiLinearMultiPickerModule } from './linear-multi-picker/linear-multi-picker.module';
import { TuiPaletteModule } from './palette/palette.module';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/hint";
import * as i2 from "@taiga-ui/core/components/data-list";
class TuiColorSelectorModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiColorSelectorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiColorSelectorModule, declarations: [TuiColorSelectorComponent], imports: [CommonModule,
            TuiButton,
            TuiSvgComponent,
            TuiColorPickerModule,
            TuiLinearMultiPickerModule,
            TuiActiveZone,
            TuiColorEditModule,
            TuiGroup,
            TuiPaletteModule,
            TuiChevron,
            TuiIcon, i1.TuiHintComponent, i1.TuiHintDirective, i1.TuiHintOptionsDirective, i1.TuiHintUnstyled, i1.TuiHintDriver, i1.TuiHintPosition, i1.TuiHintHover, i1.TuiHintOverflow, i1.TuiHintDescribe, i1.TuiHintHost, i1.TuiHintManual, i1.TuiHintPointer, i2.TuiDataListComponent, i2.TuiDataListDirective, i2.TuiOption, i2.TuiOptGroup], exports: [TuiColorSelectorComponent, i1.TuiHintComponent, i1.TuiHintDirective, i1.TuiHintOptionsDirective, i1.TuiHintUnstyled, i1.TuiHintDriver, i1.TuiHintPosition, i1.TuiHintHover, i1.TuiHintOverflow, i1.TuiHintDescribe, i1.TuiHintHost, i1.TuiHintManual, i1.TuiHintPointer, i2.TuiDataListComponent, i2.TuiDataListDirective, i2.TuiOption, i2.TuiOptGroup] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiColorSelectorModule, imports: [CommonModule,
            TuiSvgComponent,
            TuiColorPickerModule,
            TuiLinearMultiPickerModule,
            TuiColorEditModule,
            TuiPaletteModule,
            TuiIcon, i1.TuiHintComponent, i2.TuiDataListComponent, i2.TuiOption] }); }
}
export { TuiColorSelectorModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiColorSelectorModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiButton,
                        TuiSvgComponent,
                        TuiColorPickerModule,
                        TuiLinearMultiPickerModule,
                        TuiActiveZone,
                        TuiColorEditModule,
                        TuiGroup,
                        TuiPaletteModule,
                        TuiChevron,
                        TuiIcon,
                        ...TuiHint,
                        ...TuiDataList,
                    ],
                    declarations: [TuiColorSelectorComponent],
                    exports: [TuiColorSelectorComponent, ...TuiHint, ...TuiDataList],
                }]
        }] });
