// @taiga-ui/cdk/esm2022/services/theme-color.service.mjs
import { DOCUMENT } from '@angular/common';
import { inject, Injectable } from '@angular/core';
import { Meta } from '@angular/platform-browser';
import { tuiCreateTokenFromFactory } from '@taiga-ui/cdk/utils';
import * as i0 from "@angular/core";
export const TUI_THEME_COLOR = tuiCreateTokenFromFactory(() => inject(Meta).getTag('name="theme-color"')?.content ?? '');
class TuiThemeColorService {
    constructor() {
        this.current = inject(TUI_THEME_COLOR);
        this.doc = inject(DOCUMENT);
        this.meta = inject(Meta);
    }
    get color() {
        return this.current;
    }
    set color(content) {
        this.meta.updateTag({ name: 'theme-color', content });
        this.doc.documentElement.style.setProperty('--tui-theme-color', content);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiThemeColorService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiThemeColorService, providedIn: 'root' }); }
}
export { TuiThemeColorService };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiThemeColorService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
