// @ng-web-apis/intersection-observer/esm2022/directives/intersection-root.directive.mjs
import { Directive, ElementRef } from '@angular/core';
import { INTERSECTION_ROOT } from '../tokens/intersection-root';
import * as i0 from "@angular/core";
class WaIntersectionRoot {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: WaIntersectionRoot, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: WaIntersectionRoot, isStandalone: true, selector: "[waIntersectionRoot]", providers: [
            {
                provide: INTERSECTION_ROOT,
                useExisting: ElementRef,
            },
        ], ngImport: i0 });
}
export { WaIntersectionRoot };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: WaIntersectionRoot, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[waIntersectionRoot]',
                    providers: [
                        {
                            provide: INTERSECTION_ROOT,
                            useExisting: ElementRef,
                        },
                    ],
                }]
        }] });
/**
 * @deprecated: use {@link WaIntersectionRoot}
 */
export const IntersectionRootDirective = WaIntersectionRoot;
