// @ng-web-apis/intersection-observer/esm2022/directives/intersection-observee.directive.mjs
import { Directive, inject } from '@angular/core';
import { IntersectionObserveeService } from '../services/intersection-observee.service';
import * as i0 from "@angular/core";
class WaIntersectionObservee {
    waIntersectionObservee = inject(IntersectionObserveeService);
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: WaIntersectionObservee, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: WaIntersectionObservee, isStandalone: true, selector: "[waIntersectionObservee]", outputs: { waIntersectionObservee: "waIntersectionObservee" }, providers: [IntersectionObserveeService], ngImport: i0 });
}
export { WaIntersectionObservee };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: WaIntersectionObservee, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[waIntersectionObservee]',
                    outputs: ['waIntersectionObservee'],
                    providers: [IntersectionObserveeService],
                }]
        }] });
/**
 * @deprecated: use {@link WaIntersectionObservee}
 */
export const IntersectionObserveeDirective = WaIntersectionObservee;
/**
 * @deprecated: use {@link WaIntersectionObservee}
 */
export const WaObservee = WaIntersectionObservee;
