// @taiga-ui/kit/esm2022/components/push/push.directive.mjs
import { ChangeDetectorRef, Directive, forwardRef, inject, Input, TemplateRef, } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { tuiIfMap } from '@taiga-ui/cdk/observables';
import { PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import { Subject } from 'rxjs';
import { TuiPushService } from './push.service';
import * as i0 from "@angular/core";
class TuiPushDirective extends PolymorpheusTemplate {
    constructor() {
        super(inject(TemplateRef), inject(ChangeDetectorRef));
        this.push = inject(forwardRef(() => TuiPushService));
        this.show$ = new Subject();
        this.show$
            .pipe(tuiIfMap(() => this.push.open(this)), takeUntilDestroyed())
            .subscribe();
    }
    set tuiPush(show) {
        this.show$.next(show);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPushDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiPushDirective, isStandalone: true, selector: "ng-template[tuiPush]", inputs: { tuiPush: "tuiPush" }, usesInheritance: true, ngImport: i0 }); }
}
export { TuiPushDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPushDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'ng-template[tuiPush]',
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { tuiPush: [{
                type: Input
            }] } });
