// @taiga-ui/kit/esm2022/components/segmented/segmented.directive.mjs
import { ContentChildren, Directive, ElementRef, inject } from '@angular/core';
import { NgControl, RadioControlValueAccessor } from '@angular/forms';
import { RouterLinkActive } from '@angular/router';
import { EMPTY_QUERY } from '@taiga-ui/cdk/constants';
import { tuiControlValue, tuiQueryListChanges } from '@taiga-ui/cdk/observables';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { map, switchMap } from 'rxjs';
import { TuiSegmented } from './segmented.component';
import * as i0 from "@angular/core";
class TuiSegmentedDirective {
    constructor() {
        this.controls = EMPTY_QUERY;
        this.radios = EMPTY_QUERY;
        this.links = EMPTY_QUERY;
        this.elements = EMPTY_QUERY;
        this.component = inject(TuiSegmented);
        this.el = tuiInjectElement();
    }
    ngAfterContentInit() {
        tuiQueryListChanges(this.controls)
            .pipe(switchMap(() => tuiControlValue(this.controls.first)), map((value) => this.radios.toArray().findIndex((c) => c.value === value)))
            .subscribe((index) => {
            this.component.update(index);
        });
    }
    ngAfterContentChecked() {
        if (this.links.length) {
            this.update(this.elements.get(this.linkIndex)?.nativeElement || null);
        }
    }
    update(target) {
        this.component.update(this.getIndex(target));
    }
    get linkIndex() {
        return this.links.toArray().findIndex((link) => link.isActive);
    }
    getIndex(element) {
        return Array.from(this.el.children).findIndex((tab) => tab.contains(element));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSegmentedDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiSegmentedDirective, isStandalone: true, host: { listeners: { "click": "update($event.target)" } }, queries: [{ propertyName: "controls", predicate: NgControl, descendants: true }, { propertyName: "radios", predicate: RadioControlValueAccessor, descendants: true }, { propertyName: "links", predicate: RouterLinkActive }, { propertyName: "elements", predicate: RouterLinkActive, read: ElementRef }], ngImport: i0 }); }
}
export { TuiSegmentedDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSegmentedDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    host: {
                        '(click)': 'update($event.target)',
                    },
                }]
        }], propDecorators: { controls: [{
                type: ContentChildren,
                args: [NgControl, { descendants: true }]
            }], radios: [{
                type: ContentChildren,
                args: [RadioControlValueAccessor, { descendants: true }]
            }], links: [{
                type: ContentChildren,
                args: [RouterLinkActive]
            }], elements: [{
                type: ContentChildren,
                args: [RouterLinkActive, { read: ElementRef }]
            }] } });
