// @taiga-ui/legacy/esm2022/components/color-selector/color-picker/color-picker.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiFlatPickerModule } from '../flat-picker/flat-picker.module';
import { TuiLinearPickerModule } from '../linear-picker/linear-picker.module';
import { TuiColorPickerComponent } from './color-picker.component';
import * as i0 from "@angular/core";
class TuiColorPickerModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiColorPickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiColorPickerModule, declarations: [TuiColorPickerComponent], imports: [CommonModule, TuiFlatPickerModule, TuiLinearPickerModule], exports: [TuiColorPickerComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiColorPickerModule, imports: [CommonModule, TuiFlatPickerModule, TuiLinearPickerModule] }); }
}
export { TuiColorPickerModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiColorPickerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiFlatPickerModule, TuiLinearPickerModule],
                    declarations: [TuiColorPickerComponent],
                    exports: [TuiColorPickerComponent],
                }]
        }] });
