// @taiga-ui/core/esm2022/directives/appearance/with-appearance.mjs
import { Directive } from '@angular/core';
import { TuiAppearance } from './appearance.directive';
import * as i0 from "@angular/core";
import * as i1 from "./appearance.directive";
class TuiWithAppearance {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiWithAppearance, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiWithAppearance, isStandalone: true, hostDirectives: [{ directive: i1.TuiAppearance, inputs: ["tuiAppearance", "appearance", "tuiAppearanceState", "tuiAppearanceState", "tuiAppearanceFocus", "tuiAppearanceFocus", "tuiAppearanceMode", "tuiAppearanceMode"] }], ngImport: i0 }); }
}
export { TuiWithAppearance };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiWithAppearance, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    hostDirectives: [
                        {
                            directive: TuiAppearance,
                            inputs: [
                                'tuiAppearance: appearance',
                                'tuiAppearanceState',
                                'tuiAppearanceFocus',
                                'tuiAppearanceMode',
                            ],
                        },
                    ],
                }]
        }] });
