// @taiga-ui/core/esm2022/classes/accessors.mjs
import { Optional, SkipSelf } from '@angular/core';
import { tuiProvide } from '@taiga-ui/cdk/utils/miscellaneous';
export class TuiAccessor {
}
export class TuiPositionAccessor extends TuiAccessor {
}
export class TuiRectAccessor extends TuiAccessor {
}
export function tuiProvideAccessor(provide, type, fallback) {
    return {
        provide,
        deps: [[new SkipSelf(), new Optional(), provide], fallback],
        useFactory: tuiFallbackAccessor(type),
    };
}
export function tuiFallbackAccessor(type) {
    return (accessors, fallback) => accessors?.find?.((accessor) => accessor !== fallback && accessor.type === type) || fallback;
}
export function tuiPositionAccessorFor(type, fallback) {
    return tuiProvideAccessor(TuiPositionAccessor, type, fallback);
}
export function tuiRectAccessorFor(type, fallback) {
    return tuiProvideAccessor(TuiRectAccessor, type, fallback);
}
export function tuiAsPositionAccessor(accessor) {
    return tuiProvide(TuiPositionAccessor, accessor, true);
}
export function tuiAsRectAccessor(accessor) {
    return tuiProvide(TuiRectAccessor, accessor, true);
}
