// @taiga-ui/legacy/esm2022/components/input-date-time/input-date-time.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MaskitoDirective } from '@maskito/angular';
import { TuiCalendar } from '@taiga-ui/core/components/calendar';
import { TuiIcon } from '@taiga-ui/core/components/icon';
import { TuiLink } from '@taiga-ui/core/components/link';
import { TuiAppearance } from '@taiga-ui/core/directives/appearance';
import { TuiDropdown } from '@taiga-ui/core/directives/dropdown';
import { TuiPrimitiveTextfieldModule, TuiTextfieldComponent, } from '@taiga-ui/legacy/components/primitive-textfield';
import { TuiLegacyDropdownOpenMonitorDirective, TuiTextfieldControllerModule, TuiValueAccessorModule, } from '@taiga-ui/legacy/directives';
import { PolymorpheusOutlet, PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import { TuiInputDateTimeComponent } from './input-date-time.component';
import { TuiInputDateTimeDirective } from './input-date-time.directive';
import { TuiNativeDateTimeDirective } from './native-date-time/native-date-time.directive';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/dropdown";
class TuiInputDateTimeModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateTimeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateTimeModule, declarations: [TuiInputDateTimeComponent,
            TuiInputDateTimeDirective,
            TuiNativeDateTimeDirective], imports: [CommonModule,
            MaskitoDirective,
            PolymorpheusOutlet,
            PolymorpheusTemplate,
            TuiCalendar,
            TuiLink,
            TuiPrimitiveTextfieldModule,
            TuiValueAccessorModule,
            TuiTextfieldControllerModule,
            TuiLegacyDropdownOpenMonitorDirective,
            TuiIcon,
            TuiAppearance, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection], exports: [TuiInputDateTimeComponent,
            TuiInputDateTimeDirective,
            TuiTextfieldComponent, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateTimeModule, imports: [CommonModule,
            TuiCalendar,
            TuiPrimitiveTextfieldModule,
            TuiValueAccessorModule,
            TuiTextfieldControllerModule,
            TuiIcon, i1.TuiDropdownComponent] }); }
}
export { TuiInputDateTimeModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateTimeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MaskitoDirective,
                        PolymorpheusOutlet,
                        PolymorpheusTemplate,
                        TuiCalendar,
                        TuiLink,
                        TuiPrimitiveTextfieldModule,
                        TuiValueAccessorModule,
                        TuiTextfieldControllerModule,
                        TuiLegacyDropdownOpenMonitorDirective,
                        TuiIcon,
                        TuiAppearance,
                        ...TuiDropdown,
                    ],
                    declarations: [
                        TuiInputDateTimeComponent,
                        TuiInputDateTimeDirective,
                        TuiNativeDateTimeDirective,
                    ],
                    exports: [
                        TuiInputDateTimeComponent,
                        TuiInputDateTimeDirective,
                        TuiTextfieldComponent,
                        ...TuiDropdown,
                    ],
                }]
        }] });
