// @ng-web-apis/resize-observer/esm2022/directives/resize-observer.directive.mjs
import { Directive, ElementRef, inject } from '@angular/core';
import { ResizeObserverService } from '../services/resize-observer.service';
import { RESIZE_OPTION_BOX, RESIZE_OPTION_BOX_DEFAULT } from '../tokens/resize-option-box';
import * as i0 from "@angular/core";
class WaResizeObserver {
    waResizeObserver = inject(ResizeObserverService);
    box = RESIZE_OPTION_BOX_DEFAULT;
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: WaResizeObserver, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: WaResizeObserver, isStandalone: true, selector: "[waResizeObserver]", inputs: { waResizeBox: ["box", "waResizeBox"] }, outputs: { waResizeObserver: "waResizeObserver" }, providers: [
            ResizeObserverService,
            {
                provide: RESIZE_OPTION_BOX,
                useFactory: () => inject(ElementRef).nativeElement.getAttribute('waResizeBox') ||
                    RESIZE_OPTION_BOX_DEFAULT,
            },
        ], ngImport: i0 });
}
export { WaResizeObserver };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: WaResizeObserver, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[waResizeObserver]',
                    inputs: ['waResizeBox: box'],
                    outputs: ['waResizeObserver'],
                    providers: [
                        ResizeObserverService,
                        {
                            provide: RESIZE_OPTION_BOX,
                            useFactory: () => inject(ElementRef).nativeElement.getAttribute('waResizeBox') ||
                                RESIZE_OPTION_BOX_DEFAULT,
                        },
                    ],
                }]
        }] });
/**
 * @deprecated: use {@link WaResizeObserver}
 */
export const ResizeObserverDirective = WaResizeObserver;
