// @taiga-ui/kit/esm2022/components/confirm/confirm.service.mjs
import { inject, Injectable } from '@angular/core';
import { TuiDialogService } from '@taiga-ui/core/components/dialog';
import { defer, of } from 'rxjs';
import { TUI_CONFIRM } from './confirm.component';
import * as i0 from "@angular/core";
class TuiConfirmService {
    constructor() {
        this.dialogs = inject(TuiDialogService);
        this.dirty = false;
    }
    markAsDirty() {
        this.dirty = true;
    }
    markAsPristine() {
        this.dirty = false;
    }
    withConfirm(options) {
        return defer(() => this.dirty
            ? this.dialogs.open(TUI_CONFIRM, {
                size: 's',
                ...options,
            })
            : of(true));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiConfirmService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiConfirmService }); }
}
export { TuiConfirmService };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiConfirmService, decorators: [{
            type: Injectable
        }] });
