// @taiga-ui/legacy/esm2022/pipes/format-phone/format-phone.pipe.mjs
/// <reference types="@taiga-ui/tsconfig/ng-dev-mode" />
import { Pipe } from '@angular/core';
import { CHAR_PLUS } from '@taiga-ui/cdk/constants';
import { tuiFormatPhone } from '@taiga-ui/legacy/utils';
import * as i0 from "@angular/core";
/**
 * @deprecated: drop in v5.0
 */
class TuiFormatPhonePipe {
    /**
     * Use {@link https://maskito.dev/frameworks/angular#pipe MaskitoPipe} instead!
     * ```ts
     * import {maskitoTransform} from '@maskito/core';
     * import {maskitoPhoneOptionsGenerator} from '@maskito/phone';
     * import metadata from 'libphonenumber-js/min/metadata';
     *
     * @Component({
     *     standalone: true,
     *     imports: [MaskitoPipe],
     *     template: `
     *         Phone: {{'12125552368' | maskito: options}}
     *         <!-- Phone: +1 (212) 555-2368 -->
     *     `,
     * })
     * export class YourComponent {
     *     options = maskitoPhoneOptionsGenerator({countryIsoCode: 'US', metadata});
     * }
     * ```
     * ___
     * @deprecated Use {@link https://maskito.dev/frameworks/angular#pipe MaskitoPipe} instead!
     * Formats phone number string of +7XXXXXXXXXX format with adding separator symbols
     * @param value
     * @param countryCode
     * @param phoneMask
     * @return formatted phone number string +7(XXX)XXX-XX-XX
     */
    transform(value, countryCode = `${CHAR_PLUS}7`, phoneMask = '(###) ###-##-##') {
        ngDevMode &&
            console.assert(value !== undefined, 'Undefined was passed to tuiFormatPhone pipe');
        return tuiFormatPhone(value, countryCode, phoneMask);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiFormatPhonePipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiFormatPhonePipe, isStandalone: true, name: "tuiFormatPhone" }); }
}
export { TuiFormatPhonePipe };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiFormatPhonePipe, decorators: [{
            type: Pipe,
            args: [{
                    standalone: true,
                    name: 'tuiFormatPhone',
                }]
        }] });
