// @taiga-ui/legacy/esm2022/directives/textfield-controller/textfield-controller.module.mjs
import { NgModule } from '@angular/core';
import { PolymorpheusOutlet, PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import { TuiTextfieldAppearanceDirective } from './textfield-appearance.directive';
import { TuiTextfieldCleanerDirective } from './textfield-cleaner.directive';
import { TuiTextfieldCustomContentDirective } from './textfield-custom-content.directive';
import { TuiTextfieldFillerDirective } from './textfield-filler.directive';
import { TuiTextfieldIconDirective } from './textfield-icon.directive';
import { TuiTextfieldIconLeftDirective } from './textfield-icon-left.directive';
import { TuiTextfieldLabelOutsideDirective } from './textfield-label-outside.directive';
import { TuiTextfieldPostfixDirective } from './textfield-postfix.directive';
import { TuiTextfieldPrefixDirective } from './textfield-prefix.directive';
import { TuiTextfieldSizeDirective } from './textfield-size.directive';
import * as i0 from "@angular/core";
class TuiTextfieldControllerModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextfieldControllerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiTextfieldControllerModule, declarations: [TuiTextfieldAppearanceDirective,
            TuiTextfieldCleanerDirective,
            TuiTextfieldCustomContentDirective,
            TuiTextfieldLabelOutsideDirective,
            TuiTextfieldSizeDirective,
            TuiTextfieldIconDirective,
            TuiTextfieldIconLeftDirective,
            TuiTextfieldPrefixDirective,
            TuiTextfieldPostfixDirective,
            TuiTextfieldFillerDirective], imports: [PolymorpheusOutlet, PolymorpheusTemplate], exports: [TuiTextfieldAppearanceDirective,
            TuiTextfieldCleanerDirective,
            TuiTextfieldCustomContentDirective,
            TuiTextfieldLabelOutsideDirective,
            TuiTextfieldSizeDirective,
            TuiTextfieldIconDirective,
            TuiTextfieldIconLeftDirective,
            TuiTextfieldPrefixDirective,
            TuiTextfieldPostfixDirective,
            TuiTextfieldFillerDirective,
            PolymorpheusOutlet,
            PolymorpheusTemplate] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextfieldControllerModule }); }
}
export { TuiTextfieldControllerModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTextfieldControllerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [PolymorpheusOutlet, PolymorpheusTemplate],
                    declarations: [
                        TuiTextfieldAppearanceDirective,
                        TuiTextfieldCleanerDirective,
                        TuiTextfieldCustomContentDirective,
                        TuiTextfieldLabelOutsideDirective,
                        TuiTextfieldSizeDirective,
                        TuiTextfieldIconDirective,
                        TuiTextfieldIconLeftDirective,
                        TuiTextfieldPrefixDirective,
                        TuiTextfieldPostfixDirective,
                        TuiTextfieldFillerDirective,
                    ],
                    exports: [
                        TuiTextfieldAppearanceDirective,
                        TuiTextfieldCleanerDirective,
                        TuiTextfieldCustomContentDirective,
                        TuiTextfieldLabelOutsideDirective,
                        TuiTextfieldSizeDirective,
                        TuiTextfieldIconDirective,
                        TuiTextfieldIconLeftDirective,
                        TuiTextfieldPrefixDirective,
                        TuiTextfieldPostfixDirective,
                        TuiTextfieldFillerDirective,
                        PolymorpheusOutlet,
                        PolymorpheusTemplate,
                    ],
                }]
        }] });
