// @taiga-ui/kit/esm2022/components/badged-content/badged-content.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
class TuiBadgedContentDirective {
    constructor() {
        this.tuiSlot = 'top';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiBadgedContentDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiBadgedContentDirective, isStandalone: true, selector: "[tuiSlot]", inputs: { tuiSlot: "tuiSlot" }, ngImport: i0 }); }
}
export { TuiBadgedContentDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiBadgedContentDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiSlot]',
                }]
        }], propDecorators: { tuiSlot: [{
                type: Input
            }] } });
