// @taiga-ui/legacy/esm2022/components/sheet/directives/sheet-close/sheet-close.directive.mjs
import { Directive, inject, NgZone, Output } from '@angular/core';
import { WA_WINDOW } from '@ng-web-apis/common';
import { tuiIfMap, tuiTypedFromEvent, tuiZonefull } from '@taiga-ui/cdk/observables';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { tuiIsFalsy } from '@taiga-ui/cdk/utils/miscellaneous';
import { distinctUntilChanged, filter, merge, startWith } from 'rxjs';
import { TuiSheetComponent } from '../../components/sheet/sheet.component';
import { TUI_SHEET_CLOSE } from '../../components/sheet-heading/sheet-heading.component';
import { TUI_SHEET_DRAGGED, TUI_SHEET_SCROLL } from '../../sheet-tokens';
import * as i0 from "@angular/core";
class TuiSheetCloseDirective {
    constructor() {
        this.zone = inject(NgZone);
        this.dragged$ = inject(TUI_SHEET_DRAGGED);
        this.scroll$ = inject(TUI_SHEET_SCROLL);
        this.win = inject(WA_WINDOW);
        this.el = tuiInjectElement();
        this.sheet = inject(TuiSheetComponent);
        this.close = merge(tuiTypedFromEvent(this.el, TUI_SHEET_CLOSE), this.dragged$.pipe(tuiIfMap(() => this.scroll$.pipe(startWith(this.el.scrollTop)), tuiIsFalsy), filter((y) => this.sheet.item?.closeable && this.shouldClose(y)), distinctUntilChanged(), tuiZonefull(this.zone)));
    }
    shouldClose(scrollTop) {
        const height = Math.min(this.win.innerHeight, this.el.scrollHeight - this.win.innerHeight);
        const min = Math.min(height, this.sheet.stops[0] || Infinity);
        return scrollTop < min / 2;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSheetCloseDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiSheetCloseDirective, selector: "tui-sheet[close]", outputs: { close: "close" }, ngImport: i0 }); }
}
export { TuiSheetCloseDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSheetCloseDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: false,
                    selector: 'tui-sheet[close]',
                }]
        }], propDecorators: { close: [{
                type: Output
            }] } });
