// @taiga-ui/kit/esm2022/components/tree/directives/tree-controller.directive.mjs
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import { tuiIsPresent, tuiProvide } from '@taiga-ui/cdk/utils/miscellaneous';
import { TUI_TREE_ACCESSOR, TUI_TREE_CONTROLLER } from '../misc/tree.tokens';
import * as i0 from "@angular/core";
class TuiTreeControllerDirective {
    constructor() {
        this.items = new Map();
        this.fallback = true;
        this.map = new Map();
        this.toggled = new EventEmitter();
    }
    register(item, value) {
        this.items.set(item, value);
    }
    unregister(item) {
        this.items.delete(item);
    }
    isExpanded(item) {
        const value = this.items.get(item);
        return (value && this.map.get(value)) ?? this.fallback;
    }
    toggle(item) {
        const value = this.items.get(item);
        const expanded = this.isExpanded(item);
        if (!tuiIsPresent(value)) {
            return;
        }
        this.toggled.emit(value);
        this.map.set(value, !expanded);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTreeControllerDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiTreeControllerDirective, isStandalone: true, selector: "[tuiTreeController][map]", inputs: { fallback: ["tuiTreeController", "fallback"], map: "map" }, outputs: { toggled: "toggled" }, providers: [
            tuiProvide(TUI_TREE_ACCESSOR, TuiTreeControllerDirective),
            tuiProvide(TUI_TREE_CONTROLLER, TuiTreeControllerDirective),
        ], exportAs: ["tuiTreeController"], ngImport: i0 }); }
}
export { TuiTreeControllerDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTreeControllerDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiTreeController][map]',
                    providers: [
                        tuiProvide(TUI_TREE_ACCESSOR, TuiTreeControllerDirective),
                        tuiProvide(TUI_TREE_CONTROLLER, TuiTreeControllerDirective),
                    ],
                    exportAs: 'tuiTreeController',
                }]
        }], propDecorators: { fallback: [{
                type: Input,
                args: ['tuiTreeController']
            }], map: [{
                type: Input
            }], toggled: [{
                type: Output
            }] } });
