// @taiga-ui/kit/esm2022/components/tree/directives/tree-node.directive.mjs
import { Directive, inject, Input } from '@angular/core';
import { TuiTreeItem } from '../components/tree-item/tree-item.component';
import { TUI_TREE_ACCESSOR } from '../misc/tree.tokens';
import * as i0 from "@angular/core";
class TuiTreeNode {
    constructor() {
        this.component = inject(TuiTreeItem);
        this.directive = inject(TUI_TREE_ACCESSOR, {
            optional: true,
        });
    }
    set value(value) {
        this.directive?.register(this.component, value);
    }
    ngOnDestroy() {
        this.directive?.unregister(this.component);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTreeNode, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiTreeNode, isStandalone: true, selector: "tui-tree-item[tuiTreeNode]", inputs: { value: ["tuiTreeNode", "value"] }, ngImport: i0 }); }
}
export { TuiTreeNode };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTreeNode, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'tui-tree-item[tuiTreeNode]',
                }]
        }], propDecorators: { value: [{
                type: Input,
                args: ['tuiTreeNode']
            }] } });
