// @taiga-ui/legacy/esm2022/components/tag/tag.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiIcon } from '@taiga-ui/core/components/icon';
import { TuiLoader } from '@taiga-ui/core/components/loader';
import { PolymorpheusOutlet, PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import { TuiTagComponent } from './tag.component';
import * as i0 from "@angular/core";
/**
 * @deprecated: drop in v5.0 use {@link TuiChip}
 * https://taiga-ui.dev/components/chip
 */
class TuiTagModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTagModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiTagModule, declarations: [TuiTagComponent], imports: [CommonModule,
            FormsModule,
            TuiIcon,
            TuiLoader,
            PolymorpheusOutlet,
            PolymorpheusTemplate], exports: [TuiTagComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTagModule, imports: [CommonModule,
            FormsModule,
            TuiIcon,
            TuiLoader] }); }
}
export { TuiTagModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTagModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        TuiIcon,
                        TuiLoader,
                        PolymorpheusOutlet,
                        PolymorpheusTemplate,
                    ],
                    declarations: [TuiTagComponent],
                    exports: [TuiTagComponent],
                }]
        }] });
