// @taiga-ui/legacy/esm2022/components/multi-select/hide-selected.pipe.mjs
import { __decorate } from "tslib";
import { inject, Pipe } from '@angular/core';
import { tuiPure } from '@taiga-ui/cdk/utils/miscellaneous';
import { tuiIsFlat } from '@taiga-ui/kit/utils';
import { TuiMultiSelectComponent } from './multi-select.component';
import * as i0 from "@angular/core";
class TuiHideSelectedPipe {
    constructor() {
        this.component = inject(TuiMultiSelectComponent);
    }
    transform(items) {
        const { value, identityMatcher } = this.component;
        if (!items) {
            return null;
        }
        return tuiIsFlat(items)
            ? this.filter(items, value, identityMatcher)
            : this.filter2d(items, value, identityMatcher);
    }
    filter2d(items, value, matcher) {
        return items.map((subItems) => this.filter(subItems, value, matcher));
    }
    filter(items, value, matcher) {
        return items.filter((item) => value.every((selected) => !matcher(selected, item)));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiHideSelectedPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiHideSelectedPipe, name: "tuiHideSelected", pure: false }); }
}
__decorate([
    tuiPure
], TuiHideSelectedPipe.prototype, "filter2d", null);
__decorate([
    tuiPure
], TuiHideSelectedPipe.prototype, "filter", null);
export { TuiHideSelectedPipe };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiHideSelectedPipe, decorators: [{
            type: Pipe,
            args: [{
                    standalone: false,
                    name: 'tuiHideSelected',
                    pure: false,
                }]
        }], propDecorators: { filter2d: [], filter: [] } });
