// @taiga-ui/legacy/esm2022/components/multi-select/multi-select.directive.mjs
import { Directive } from '@angular/core';
import { AbstractTuiTextfieldHost } from '@taiga-ui/legacy/classes';
import { tuiAsTextfieldHost } from '@taiga-ui/legacy/tokens';
import * as i0 from "@angular/core";
class TuiMultiSelectDirective extends AbstractTuiTextfieldHost {
    constructor() {
        super(...arguments);
        this.disableItemHandler = (item) => this.host.disabledItemHandler(item);
    }
    get readOnly() {
        return true;
    }
    get stringify() {
        return this.host.stringify;
    }
    onValueChange() { }
    onSelectionChange(value) {
        this.host.onValueChange(value);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiMultiSelectDirective, selector: "tui-multi-select", providers: [tuiAsTextfieldHost(TuiMultiSelectDirective)], usesInheritance: true, ngImport: i0 }); }
}
export { TuiMultiSelectDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: false,
                    selector: 'tui-multi-select',
                    providers: [tuiAsTextfieldHost(TuiMultiSelectDirective)],
                }]
        }] });
