// @taiga-ui/legacy/esm2022/components/multi-select/multi-select.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiLet } from '@taiga-ui/cdk/directives/let';
import { TuiMapperPipe } from '@taiga-ui/cdk/pipes/mapper';
import { TuiDataList } from '@taiga-ui/core/components/data-list';
import { TuiLink } from '@taiga-ui/core/components/link';
import { TuiDropdown } from '@taiga-ui/core/directives/dropdown';
import { TuiDataListWrapper } from '@taiga-ui/kit/components/data-list-wrapper';
import { TuiStringifyContentPipe } from '@taiga-ui/kit/pipes/stringify-content';
import { TuiArrowComponent } from '@taiga-ui/legacy/components/arrow';
import { TuiInputTagModule } from '@taiga-ui/legacy/components/input-tag';
import { TuiMultiSelectOptionModule } from '@taiga-ui/legacy/components/multi-select-option';
import { TuiLegacyDropdownOpenMonitorDirective, TuiTextfieldControllerModule, TuiWrapperModule, } from '@taiga-ui/legacy/directives';
import { PolymorpheusOutlet, PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import { TuiHideSelectedPipe } from './hide-selected.pipe';
import { TuiMultiSelectComponent } from './multi-select.component';
import { TuiMultiSelectDirective } from './multi-select.directive';
import { TuiMultiSelectGroupComponent } from './multi-select-group/multi-select-group.component';
import { TuiMultiSelectGroupDirective } from './multi-select-group/multi-select-group.directive';
import { TuiNativeMultiSelectComponent } from './native-multi-select/native-multi-select.component';
import { TuiNativeMultiSelectGroupComponent } from './native-multi-select/native-multi-select-group.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/dropdown";
import * as i2 from "@taiga-ui/core/components/data-list";
import * as i3 from "@taiga-ui/kit/components/data-list-wrapper";
class TuiMultiSelectModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectModule, declarations: [TuiMultiSelectComponent,
            TuiMultiSelectGroupComponent,
            TuiMultiSelectGroupDirective,
            TuiHideSelectedPipe,
            TuiNativeMultiSelectComponent,
            TuiNativeMultiSelectGroupComponent,
            TuiMultiSelectDirective], imports: [CommonModule,
            FormsModule,
            PolymorpheusOutlet,
            PolymorpheusTemplate,
            TuiWrapperModule,
            TuiArrowComponent,
            TuiLet,
            TuiMapperPipe,
            TuiInputTagModule,
            TuiMultiSelectOptionModule,
            TuiMapperPipe,
            TuiLink,
            TuiTextfieldControllerModule,
            TuiStringifyContentPipe,
            TuiLegacyDropdownOpenMonitorDirective, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection, i2.TuiDataListComponent, i2.TuiDataListDirective, i2.TuiOption, i2.TuiOptGroup, i3.TuiDataListWrapperComponent, i3.TuiDataListGroupWrapperComponent, i2.TuiDataListDirective], exports: [TuiMultiSelectComponent,
            TuiMultiSelectGroupComponent,
            TuiMultiSelectGroupDirective,
            TuiHideSelectedPipe,
            TuiMultiSelectDirective,
            TuiNativeMultiSelectComponent,
            TuiNativeMultiSelectGroupComponent, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection, i2.TuiDataListComponent, i2.TuiDataListDirective, i2.TuiOption, i2.TuiOptGroup, i3.TuiDataListWrapperComponent, i3.TuiDataListGroupWrapperComponent, i2.TuiDataListDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectModule, imports: [CommonModule,
            FormsModule,
            TuiWrapperModule,
            TuiArrowComponent,
            TuiInputTagModule,
            TuiMultiSelectOptionModule,
            TuiTextfieldControllerModule, i1.TuiDropdownComponent, i2.TuiDataListComponent, i2.TuiOption, i3.TuiDataListWrapperComponent, i3.TuiDataListGroupWrapperComponent] }); }
}
export { TuiMultiSelectModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        PolymorpheusOutlet,
                        PolymorpheusTemplate,
                        TuiWrapperModule,
                        TuiArrowComponent,
                        TuiLet,
                        TuiMapperPipe,
                        TuiInputTagModule,
                        TuiMultiSelectOptionModule,
                        TuiMapperPipe,
                        TuiLink,
                        TuiTextfieldControllerModule,
                        TuiStringifyContentPipe,
                        TuiLegacyDropdownOpenMonitorDirective,
                        ...TuiDropdown,
                        ...TuiDataList,
                        ...TuiDataListWrapper,
                    ],
                    declarations: [
                        TuiMultiSelectComponent,
                        TuiMultiSelectGroupComponent,
                        TuiMultiSelectGroupDirective,
                        TuiHideSelectedPipe,
                        TuiNativeMultiSelectComponent,
                        TuiNativeMultiSelectGroupComponent,
                        TuiMultiSelectDirective,
                    ],
                    exports: [
                        TuiMultiSelectComponent,
                        TuiMultiSelectGroupComponent,
                        TuiMultiSelectGroupDirective,
                        TuiHideSelectedPipe,
                        TuiMultiSelectDirective,
                        TuiNativeMultiSelectComponent,
                        TuiNativeMultiSelectGroupComponent,
                        ...TuiDropdown,
                        ...TuiDataList,
                        ...TuiDataListWrapper,
                    ],
                }]
        }] });
