// @taiga-ui/core/esm2022/directives/dropdown/with-dropdown-open.directive.mjs
import { Directive } from '@angular/core';
import { TuiDropdownOpen } from './dropdown-open.directive';
import * as i0 from "@angular/core";
import * as i1 from "./dropdown-open.directive";
class TuiWithDropdownOpen {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiWithDropdownOpen, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiWithDropdownOpen, isStandalone: true, hostDirectives: [{ directive: i1.TuiDropdownOpen, inputs: ["tuiDropdownOpen", "open"], outputs: ["tuiDropdownOpenChange", "openChange"] }], ngImport: i0 }); }
}
export { TuiWithDropdownOpen };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiWithDropdownOpen, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    hostDirectives: [
                        {
                            directive: TuiDropdownOpen,
                            inputs: ['tuiDropdownOpen: open'],
                            outputs: ['tuiDropdownOpenChange: openChange'],
                        },
                    ],
                }]
        }] });
