// @taiga-ui/addon-mobile/esm2022/components/mobile-calendar/mobile-calendar.providers.mjs
import { VIRTUAL_SCROLL_STRATEGY } from '@angular/cdk/scrolling';
import { ChangeDetectorRef, InjectionToken, Optional } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { tuiWatch } from '@taiga-ui/cdk/observables';
import { TuiScrollService } from '@taiga-ui/cdk/services';
import { TUI_IS_IOS } from '@taiga-ui/cdk/tokens';
import { TUI_CALENDAR_DATE_STREAM } from '@taiga-ui/kit/tokens';
import { EMPTY } from 'rxjs';
import { TuiMobileCalendarStrategy } from './mobile-calendar.strategy';
/**
 * Stream for updating value
 */
export const TUI_VALUE_STREAM = new InjectionToken('[TUI_VALUE_STREAM]');
export const TUI_MOBILE_CALENDAR_PROVIDERS = [
    TuiScrollService,
    {
        provide: VIRTUAL_SCROLL_STRATEGY,
        deps: [TUI_IS_IOS, TuiScrollService],
        useClass: TuiMobileCalendarStrategy,
    },
    {
        provide: TUI_VALUE_STREAM,
        deps: [[new Optional(), TUI_CALENDAR_DATE_STREAM], ChangeDetectorRef],
        useFactory: (value$, cdr) => (value$ || EMPTY).pipe(tuiWatch(cdr), takeUntilDestroyed()),
    },
];
