// @taiga-ui/kit/esm2022/components/preview/preview-action/preview-action.directive.mjs
import { Directive } from '@angular/core';
import { tuiButtonOptionsProvider } from '@taiga-ui/core/components/button';
import * as i0 from "@angular/core";
class TuiPreviewAction {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPreviewAction, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiPreviewAction, isStandalone: true, selector: "[tuiPreviewAction]", host: { properties: { "style.border-radius.rem": "100" } }, providers: [
            tuiButtonOptionsProvider({
                appearance: 'preview-action',
                size: 's',
            }),
        ], ngImport: i0 }); }
}
export { TuiPreviewAction };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiPreviewAction, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiPreviewAction]',
                    providers: [
                        tuiButtonOptionsProvider({
                            appearance: 'preview-action',
                            size: 's',
                        }),
                    ],
                    host: {
                        '[style.border-radius.rem]': '100',
                    },
                }]
        }] });
