// @taiga-ui/kit/esm2022/components/radio/radio.directive.mjs
import { Directive, inject, Input } from '@angular/core';
import { RadioControlValueAccessor } from '@angular/forms';
import { TUI_DEFAULT_IDENTITY_MATCHER } from '@taiga-ui/cdk/constants';
import * as i0 from "@angular/core";
class TuiRadioDirective {
    constructor() {
        this.identityMatcher = TUI_DEFAULT_IDENTITY_MATCHER;
        const accessor = inject(RadioControlValueAccessor);
        const writeValue = accessor.writeValue.bind(accessor);
        accessor.writeValue = (value) => {
            if (this.identityMatcher(value, accessor.value)) {
                writeValue(accessor.value);
            }
            else {
                writeValue(value);
            }
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiRadioDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiRadioDirective, isStandalone: true, selector: "input[type=\"radio\"][tuiRadio][identityMatcher]", inputs: { identityMatcher: "identityMatcher" }, ngImport: i0 }); }
}
export { TuiRadioDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiRadioDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'input[type="radio"][tuiRadio][identityMatcher]',
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { identityMatcher: [{
                type: Input
            }] } });
