// @taiga-ui/cdk/esm2022/directives/hovered/hovered.directive.mjs
import { Directive, inject, Output } from '@angular/core';
import { TuiHoveredService } from './hovered.service';
import * as i0 from "@angular/core";
class TuiHovered {
    constructor() {
        this.tuiHoveredChange = inject(TuiHoveredService);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiHovered, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiHovered, isStandalone: true, selector: "[tuiHoveredChange]", outputs: { tuiHoveredChange: "tuiHoveredChange" }, providers: [TuiHoveredService], ngImport: i0 }); }
}
export { TuiHovered };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiHovered, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiHoveredChange]',
                    providers: [TuiHoveredService],
                }]
        }], propDecorators: { tuiHoveredChange: [{
                type: Output
            }] } });
