// @taiga-ui/legacy/esm2022/components/input-tag/input-tag.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiHovered } from '@taiga-ui/cdk/directives/hovered';
import { TuiScrollService } from '@taiga-ui/cdk/services';
import { TuiIcon } from '@taiga-ui/core/components/icon';
import { TuiScrollbar } from '@taiga-ui/core/components/scrollbar';
import { TuiAppearance } from '@taiga-ui/core/directives/appearance';
import { TuiDropdown } from '@taiga-ui/core/directives/dropdown';
import { TuiTagModule } from '@taiga-ui/legacy/components/tag';
import { TuiTooltipModule } from '@taiga-ui/legacy/components/tooltip';
import { TuiLegacyDropdownOpenMonitorDirective, TuiWrapperModule, } from '@taiga-ui/legacy/directives';
import { PolymorpheusOutlet, PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import { TuiInputTagComponent } from './input-tag.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/directives/dropdown";
class TuiInputTagModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputTagModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiInputTagModule, declarations: [TuiInputTagComponent], imports: [CommonModule,
            FormsModule,
            PolymorpheusOutlet,
            PolymorpheusTemplate,
            TuiHovered,
            TuiScrollbar,
            TuiTooltipModule,
            TuiTagModule,
            TuiWrapperModule,
            TuiLegacyDropdownOpenMonitorDirective,
            TuiAppearance,
            TuiIcon, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection], exports: [TuiInputTagComponent, i1.TuiDropdownOptionsDirective, i1.TuiDropdownDriverDirective, i1.TuiDropdownDirective, i1.TuiDropdownComponent, i1.TuiDropdownOpen, i1.TuiDropdownOpenLegacy, i1.TuiDropdownPortal, i1.TuiDropdownManual, i1.TuiDropdownHover, i1.TuiDropdownContext, i1.TuiDropdownPosition, i1.TuiDropdownPositionSided, i1.TuiDropdownSelection] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputTagModule, providers: [TuiScrollService], imports: [CommonModule,
            FormsModule,
            TuiScrollbar,
            TuiTooltipModule,
            TuiTagModule,
            TuiWrapperModule,
            TuiIcon, i1.TuiDropdownComponent] }); }
}
export { TuiInputTagModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputTagModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        PolymorpheusOutlet,
                        PolymorpheusTemplate,
                        TuiHovered,
                        TuiScrollbar,
                        TuiTooltipModule,
                        TuiTagModule,
                        TuiWrapperModule,
                        TuiLegacyDropdownOpenMonitorDirective,
                        TuiAppearance,
                        TuiIcon,
                        ...TuiDropdown,
                    ],
                    declarations: [TuiInputTagComponent],
                    providers: [TuiScrollService],
                    exports: [TuiInputTagComponent, ...TuiDropdown],
                }]
        }] });
